{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "24795c2cd4018e77",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-07T00:23:15.945569Z",
     "start_time": "2025-01-07T00:23:15.942500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "from collections import defaultdict\n",
    "import gffutils\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "95c69da24d52a1b5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-07T00:23:17.247746Z",
     "start_time": "2025-01-07T00:23:16.394375Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# -----------------------------------------------------------------------------\n",
    "# 1. Read the BED12 file into a pandas DataFrame\n",
    "# -----------------------------------------------------------------------------\n",
    "bed_file = \"/Data1/emmab/software/code/data/L1/combined/L1_non_stringent2_full_length.bed\"\n",
    "cols = [\n",
    "    \"chrom\",        # 0\n",
    "    \"chromStart\",   # 1\n",
    "    \"chromEnd\",     # 2\n",
    "    \"name\",         # 3 (read ID)\n",
    "    \"score\",        # 4\n",
    "    \"strand\",       # 5\n",
    "    \"thickStart\",   # 6\n",
    "    \"thickEnd\",     # 7\n",
    "    \"itemRgb\",      # 8\n",
    "    \"blockCount\",   # 9\n",
    "    \"blockSizes\",   # 10\n",
    "    \"blockStarts\"   # 11\n",
    "]\n",
    "df_bed = pd.read_csv(bed_file, sep=\"\\t\", header=None, names=cols)\n",
    "df_bed.head()\n",
    "\n",
    "# keep only first 10,000 rows\n",
    "df_bed = df_bed.head(10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bb103dce6274e538",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-07T00:23:18.774205Z",
     "start_time": "2025-01-07T00:23:18.257180Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>read_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrIII</td>\n",
       "      <td>7855133</td>\n",
       "      <td>7855380</td>\n",
       "      <td>+</td>\n",
       "      <td>15de427b-8540-4ce5-b0c7-42b46d5539df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrIII</td>\n",
       "      <td>7855433</td>\n",
       "      <td>7855646</td>\n",
       "      <td>+</td>\n",
       "      <td>15de427b-8540-4ce5-b0c7-42b46d5539df</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2876929</td>\n",
       "      <td>2877086</td>\n",
       "      <td>+</td>\n",
       "      <td>631fedea-2125-4c77-9be6-534e82449221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2877136</td>\n",
       "      <td>2877352</td>\n",
       "      <td>+</td>\n",
       "      <td>631fedea-2125-4c77-9be6-534e82449221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2877403</td>\n",
       "      <td>2877715</td>\n",
       "      <td>+</td>\n",
       "      <td>631fedea-2125-4c77-9be6-534e82449221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end strand                               read_id\n",
       "0  chrIII  7855133  7855380      +  15de427b-8540-4ce5-b0c7-42b46d5539df\n",
       "1  chrIII  7855433  7855646      +  15de427b-8540-4ce5-b0c7-42b46d5539df\n",
       "2    chrI  2876929  2877086      +  631fedea-2125-4c77-9be6-534e82449221\n",
       "3    chrI  2877136  2877352      +  631fedea-2125-4c77-9be6-534e82449221\n",
       "4    chrI  2877403  2877715      +  631fedea-2125-4c77-9be6-534e82449221"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 2. Expand BED12 reads into individual exons.\n",
    "#    Each BED12 entry can have multiple exons (\"blocks\").\n",
    "# -----------------------------------------------------------------------------\n",
    "def expand_bed12(df):\n",
    "    \"\"\"Return a DataFrame of exons from BED12 entries.\"\"\"\n",
    "    # We'll store exon-level entries in a list of dicts, then convert to DF.\n",
    "    exon_entries = []\n",
    "    for idx, row in df.iterrows():\n",
    "        chrom = row[\"chrom\"]\n",
    "        strand = row[\"strand\"]\n",
    "        start = row[\"chromStart\"]\n",
    "        block_count = row[\"blockCount\"]\n",
    "        block_sizes = list(map(int, row[\"blockSizes\"].split(\",\")))\n",
    "        block_starts = list(map(int, row[\"blockStarts\"].split(\",\")))\n",
    "        read_id = row[\"name\"]\n",
    "        \n",
    "        for i in range(block_count):\n",
    "            exon_start = start + block_starts[i]\n",
    "            exon_end = exon_start + block_sizes[i]\n",
    "            exon_entries.append({\n",
    "                \"chrom\": chrom,\n",
    "                \"start\": exon_start,\n",
    "                \"end\": exon_end,\n",
    "                \"strand\": strand,\n",
    "                \"read_id\": read_id\n",
    "            })\n",
    "    return pd.DataFrame(exon_entries)\n",
    "\n",
    "exons_bed = expand_bed12(df_bed)\n",
    "exons_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7e784b9fc1d7dfdc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-07T00:23:22.505821Z",
     "start_time": "2025-01-07T00:23:19.707882Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform_key</th>\n",
       "      <th>read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(chrI, -, ((2069085, 2069293), (2069342, 20695...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(chrM, +, ((11366, 11691),))</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(chrI, +, ((4790285, 4790341), (4790412, 47906...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(chrV, -, ((12727633, 12727859), (12727905, 12...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(chrI, -, ((6220089, 6220205), (6220367, 62205...</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         isoform_key  read_count\n",
       "0  (chrI, -, ((2069085, 2069293), (2069342, 20695...          17\n",
       "1                       (chrM, +, ((11366, 11691),))          14\n",
       "2  (chrI, +, ((4790285, 4790341), (4790412, 47906...          13\n",
       "3  (chrV, -, ((12727633, 12727859), (12727905, 12...          12\n",
       "4  (chrI, -, ((6220089, 6220205), (6220367, 62205...          12"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 3. Group exons by (chrom, strand, blockCount, positions) to define isoforms.\n",
    "#    We'll define a \"canonical\" isoform key by sorting exons and recording\n",
    "#    their start-end intervals.\n",
    "# -----------------------------------------------------------------------------\n",
    "def build_isoform_key(group):\n",
    "    \"\"\"Build a tuple of (chrom, strand, [(exon_start, exon_end), ...]) sorted by start.\"\"\"\n",
    "    # group is all exons for a single read\n",
    "    chrom = group.iloc[0][\"chrom\"]\n",
    "    strand = group.iloc[0][\"strand\"]\n",
    "    \n",
    "    # Sort exons by start, then build a list of (start, end)\n",
    "    exon_coords = sorted(zip(group[\"start\"], group[\"end\"]), key=lambda x: x[0])\n",
    "    return (chrom, strand, tuple(exon_coords))\n",
    "\n",
    "# Map each read_id -> isoform_key\n",
    "read_to_isoform = {}\n",
    "for read_id, grp in exons_bed.groupby(\"read_id\"):\n",
    "    iso_key = build_isoform_key(grp)\n",
    "    read_to_isoform[read_id] = iso_key\n",
    "\n",
    "# Assign isoform_key to each row in df_bed\n",
    "df_bed[\"isoform_key\"] = df_bed[\"name\"].apply(lambda rid: read_to_isoform[rid])\n",
    "\n",
    "# Count how many reads per isoform\n",
    "isoform_counts = (\n",
    "    df_bed.groupby(\"isoform_key\")\n",
    "    .size()\n",
    "    .reset_index(name=\"read_count\")\n",
    "    .sort_values(\"read_count\", ascending=False)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "isoform_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a30a88336970cfe3",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# wget https://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS19/species/caenorhabditis_elegans/PRJNA13758/caenorhabditis_elegans.PRJNA13758.WBPS19.annotations.gff3.gz\n",
    "# \n",
    "# import gzip\n",
    "# import shutil\n",
    "# import os\n",
    "# \n",
    "# # Path to the gzipped file\n",
    "# gzipped_file = \"/Data1/reference/caenorhabditis_elegans.PRJNA13758.WBPS19.annotations.gff3.gz\"\n",
    "# output_file = gzipped_file.rstrip('.gz')  # Remove .gz extension for the output file\n",
    "# \n",
    "# # Gunzip the file\n",
    "# with gzip.open(gzipped_file, 'rb') as f_in:\n",
    "#     with open(output_file, 'wb') as f_out:\n",
    "#         shutil.copyfileobj(f_in, f_out)\n",
    "# \n",
    "# # Delete the original gzipped file\n",
    "# os.remove(gzipped_file)\n",
    "# \n",
    "# print(f\"Decompressed and removed: {gzipped_file}\")\n",
    "# print(f\"Output file created: {output_file}\")\n",
    "\n",
    "# pass cell\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59390c7e44900114",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2025-01-07T00:23:29.022072Z"
    },
    "collapsed": false,
    "is_executing": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# -----------------------------------------------------------------------------\n",
    "# 4. Load reference GFF3 annotation using gffutils\n",
    "# -----------------------------------------------------------------------------\n",
    "gff_file = \"/Data1/reference/caenorhabditis_elegans.PRJNA13758.WBPS19.annotations.gff3\"\n",
    "\n",
    "# Create or connect to the database\n",
    "db_file = 'c_elegans.db'\n",
    "\n",
    "# Check if the database already exists\n",
    "if os.path.exists(db_file):\n",
    "    print(f\"Loading existing gffutils database from '{db_file}'...\")\n",
    "    db = gffutils.FeatureDB(db_file, keep_order=True)\n",
    "else:\n",
    "    print(f\"Creating gffutils database from '{gff_file}'...\")\n",
    "    db = gffutils.create_db(\n",
    "        gff_file,\n",
    "        dbfn=db_file,\n",
    "        force=True,\n",
    "        keep_order=True,\n",
    "        merge_strategy='merge',\n",
    "        sort_attribute_values=True\n",
    "    )\n",
    "    print(f\"Database created and saved to '{db_file}'.\")"
   ]
  },
  {
   "cell_type": "code",
   "id": "5b3b2546346bc053",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Extract all gene features\n",
    "gene_features = list(db.features_of_type('gene'))\n",
    "\n",
    "# Build a mapping from gene ID to gene symbol\n",
    "gene_id_to_symbol = {}\n",
    "for gene in gene_features:\n",
    "    gene_id = gene.id\n",
    "    # Extract gene symbol from attributes; adjust the key if necessary\n",
    "    # Common attribute keys: 'Name', 'gene', 'gene_symbol'\n",
    "    gene_symbol = gene.attributes.get('Name', [gene.id])[0]  # Fallback to gene ID if Name not present\n",
    "    gene_id_to_symbol[gene_id] = gene_symbol\n",
    "\n",
    "# Extract all transcript features\n",
    "transcript_features = list(db.features_of_type('transcript'))\n",
    "\n",
    "# Build a mapping from transcript ID to gene ID and gene symbol\n",
    "transcript_to_gene = {}\n",
    "for transcript in transcript_features:\n",
    "    transcript_id = transcript.id\n",
    "    # Extract Parent gene ID from attributes; adjust the key if necessary\n",
    "    # Common attribute keys: 'Parent', 'gene_id', 'gene'\n",
    "    parent_gene_ids = transcript.attributes.get('Parent', [])\n",
    "    if not parent_gene_ids:\n",
    "        # Some GFF3 files might use 'gene' or 'gene_id' instead\n",
    "        parent_gene_ids = transcript.attributes.get('gene', [])\n",
    "    if not parent_gene_ids:\n",
    "        # If still not found, use the transcript ID as gene ID (unlikely)\n",
    "        parent_gene_ids = [transcript_id]\n",
    "    # Assuming one parent gene per transcript\n",
    "    parent_gene_id = parent_gene_ids[0]\n",
    "    gene_symbol = gene_id_to_symbol.get(parent_gene_id, parent_gene_id)\n",
    "    transcript_to_gene[transcript_id] = {\n",
    "        \"gene_id\": parent_gene_id,\n",
    "        \"gene_symbol\": gene_symbol\n",
    "    }\n",
    "\n",
    "transcript_exons = defaultdict(list)\n",
    "for transcript in transcript_features:\n",
    "    transcript_id = transcript.id\n",
    "    gene_info = transcript_to_gene.get(transcript_id, {\"gene_id\": \"unknown\", \"gene_symbol\": \"unknown\"})\n",
    "    exons = list(db.children(transcript, featuretype='exon', order_by='start'))\n",
    "    exon_coords = sorted([(e.start, e.end) for e in exons], key=lambda x: x[0])\n",
    "    key = (transcript.chrom, transcript.strand, tuple(exon_coords))\n",
    "    # Store transcript ID and gene symbol\n",
    "    transcript_exons[key].append({\n",
    "        \"transcript_id\": transcript_id,\n",
    "        \"gene_symbol\": gene_info[\"gene_symbol\"]\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e39d23bf50c1e4b6",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def classify_isoform(iso_key):\n",
    "    \"\"\"Return a dict with classification (known or novel) \n",
    "       and possible matching transcript IDs and gene symbols.\"\"\"\n",
    "    if iso_key in transcript_exons:\n",
    "        matching_info = transcript_exons[iso_key]\n",
    "        transcript_ids = [info[\"transcript_id\"] for info in matching_info]\n",
    "        gene_symbols = list(set(info[\"gene_symbol\"] for info in matching_info))\n",
    "        return {\n",
    "            \"classification\": \"known\",\n",
    "            \"matching_transcripts\": \";\".join(transcript_ids),\n",
    "            \"gene_symbols\": \";\".join(gene_symbols)\n",
    "        }\n",
    "    else:\n",
    "        return {\n",
    "            \"classification\": \"novel\",\n",
    "            \"matching_transcripts\": \"\",\n",
    "            \"gene_symbols\": \"\"\n",
    "        }\n",
    "\n",
    "classification_results = []\n",
    "for i, row in isoform_counts.iterrows():\n",
    "    iso_key = row[\"isoform_key\"]\n",
    "    read_count = row[\"read_count\"]\n",
    "    class_info = classify_isoform(iso_key)\n",
    "    classification_results.append({\n",
    "        \"isoform_key\": iso_key,\n",
    "        \"read_count\": read_count,\n",
    "        \"classification\": class_info[\"classification\"],\n",
    "        \"matching_transcripts\": class_info[\"matching_transcripts\"],\n",
    "        \"gene_symbols\": class_info[\"gene_symbols\"]\n",
    "    })\n",
    "\n",
    "df_classification = pd.DataFrame(classification_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "id": "76b68cbcbd6f11de",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Show first 10 isoforms\n",
    "display(df_classification.head(10))\n",
    "\n",
    "# Save the table as CSV if desired\n",
    "output_csv = \"isoform_classification_results_with_genes.csv\"\n",
    "df_classification.to_csv(output_csv, index=False)\n",
    "print(f\"Classification results with gene symbols saved to '{output_csv}'.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
