{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bcfb240",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code that reads in nucleosome positions and performs NRL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472b0a40",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T00:29:50.243738400Z",
     "start_time": "2023-08-16T00:29:50.218739200Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import nanotools\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "from sklearn.linear_model import LinearRegression # for linear regression analysis\n",
    "from scipy.signal import find_peaks # For finding peaks in distribtion\n",
    "#sns.set_style('darkgrid')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7fbb8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "modbam2bed --extended -m 6mA -r CHROMOSOME_X:10000000-10100000 -t 6 --threshold 0.5 /Data1/reference/c_elegans.WS235.genomic.fa /Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/combined_pod5/barcode07/basecalls/barcode07.mod_mappings.sorted.bam /Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/combined_pod5/barcode08/basecalls/barcode08.mod_mappings.sorted.bam /Data1/seq_data/TubeAB_FiberSeq_TimeC_N2_021Aux_4_10_23/basecalls/m6A_full/demux/mod_mappings_barcode_05.bam /Data1/seq_data/Tube4_b2_2uM-Hia5_fiber-seq_11_21_22/basecalls/mod_mappings.sorted.m6Aonly.bam /Data1/seq_data/TubeAD1_N2_fiberseq_6_13_23/mod_basecalls/mod_mappings.sorted.bam > /Data1/seq_data/combined_N2_fiber/combined_N2_fiber.chrX100kb.m6A.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b8f40c8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T00:09:56.517024Z",
     "start_time": "2023-08-16T00:09:56.079309900Z"
    }
   },
   "outputs": [],
   "source": [
    "### Import dataframes\n",
    "# Import nucleosome positions\n",
    "# Nucleosome cutoff = 9\n",
    "nucleosome_positions_D = pd.read_csv(\"/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/wig/output/result_m6A_9/pooled/Data1_seq_data_TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22_wig_output_D_modified_bases.6mA.merged_strand_Hnorm_10SCALE.smooth.positions.xls\", delimiter=\"\\t\")\n",
    "nucleosome_positions_H = pd.read_csv(\"/Data1/seq_data/TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19/wig/output/result_m6A_9/pooled/Data1_seq_data_TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19_wig_output_H_modified_bases.6mA.merged_strand_NOnorm_10SCALE.smooth.positions.xls\", delimiter=\"\\t\")\n",
    "\n",
    "# Nucleosome cutoff = 10\n",
    "#nucleosome_positions_D = pd.read_csv(\"/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/wig/output/result_m6a_10/pooled/Data1_seq_data_TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22_wig_output_D_modified_bases.6mA.merged_strand_Hnorm_10SCALE.smooth.positions.xls\", delimiter=\"\\t\")\n",
    "#nucleosome_positions_H = pd.read_csv(\"/Data1/seq_data/TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19/wig/output/result_m6a_10/pooled/Data1_seq_data_TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19_wig_output_H_modified_bases.6mA.merged_strand_NOnorm_10SCALE.smooth.positions.xls\", delimiter=\"\\t\")\n",
    "display(nucleosome_positions_D.head(1))\n",
    "\n",
    "# Import bed file\n",
    "bed_raw = pd.read_csv(\"/Data1/reference/tss_tes_rex_combined.bed\", delimiter=\"\\t\")\n",
    "bed_raw = bed_raw.drop(columns=[\"chr-type\"])\n",
    "display('bed_raw:',bed_raw.head(1))\n",
    "#display(bed_raw.type.unique())\n",
    "bed_tss = bed_raw[bed_raw['type'].str.startswith('TSS')]\n",
    "display('bed_tss:',bed_tss.head(1))\n",
    "\n",
    "bed_rex = bed_raw[bed_raw['type'].str.endswith('rex')]\n",
    "display('bed_rex:',bed_rex.head(1))\n",
    "display('nucleosome_positions_D:',nucleosome_positions_D.head(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7d5ce3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T00:22:39.771042400Z",
     "start_time": "2023-08-16T00:22:12.665876500Z"
    }
   },
   "outputs": [],
   "source": [
    "### Compute Frequency Tables\n",
    "importlib.reload(nanotools)\n",
    "\n",
    "frequency_df_D = nanotools.frequency_table(nucleosome_positions_D)\n",
    "frequency_df_H = nanotools.frequency_table(nucleosome_positions_H)\n",
    "\n",
    "display(frequency_df_D.tail(10))     \n",
    "display(frequency_df_H.head(10))\n",
    "\n",
    "#frequency_df_D.to_csv(\"/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/wig/output/result_m6a_10/pooled/D_nrl.csv\",index=False,)\n",
    "#frequency_df_H = frequency_table(nucleosome_positions_H)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ef7fbf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-15T17:21:35.759068400Z",
     "start_time": "2023-08-15T17:21:33.467332Z"
    }
   },
   "outputs": [],
   "source": [
    "### Create data subsets for plotting\n",
    "frequency_df_D['genotype'] = 'N2'\n",
    "frequency_df_H['genotype'] = 'AID-SDC2 + Aux'\n",
    "data = pd.concat([frequency_df_D, frequency_df_H])\n",
    "data['chr-type'] = data['chr'].apply(lambda x: 'X' if x == 'CHROMOSOME_X' else 'A')\n",
    "\n",
    "data_x=data[data['chr']=='CHROMOSOME_X']\n",
    "data_a=data[data['chr']!='CHROMOSOME_X']\n",
    "data_D=data[data['genotype']=='N2']\n",
    "data_H=data[data['genotype']!='N2']\n",
    "\n",
    "#fig = sns.displot(data_D, x=\"dist\", hue=\"chr-type\", kde=True, cut=0)\n",
    "data_D_crop=data_D[data_D['dist']<2200]\n",
    "data_H_crop=data_H[data_H['dist']<2200]\n",
    "\n",
    "### Histograms, break out into X and A for separate plotting\n",
    "D_X = data_D_crop.loc[data_D_crop['chr-type']=='X']\n",
    "D_A = data_D_crop.loc[data_D_crop['chr-type']=='A']\n",
    "H_X = data_H_crop.loc[data_H_crop['chr-type']=='X']\n",
    "H_A = data_H_crop.loc[data_H_crop['chr-type']=='A']\n",
    "\n",
    "print(D_A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ec23a1d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:42:27.857211700Z",
     "start_time": "2023-08-16T01:42:04.737319700Z"
    }
   },
   "outputs": [],
   "source": [
    "### PLOT NRL DISTRIBUTION BY CHROMOSOME FOR TWO CONDITIONS (N2 and #021)\n",
    "importlib.reload(nanotools)\n",
    "\n",
    "size = (8,12)\n",
    "\n",
    "# Assuming D_X, D_A, etc., are your dataframes for chromosomes\n",
    "# You will need to filter D_A for each chromosome I to V\n",
    "chromosome_data = [\n",
    "    (D_X, \"CHROMOSOME_X\", \"#c45746\", \"N2_30min_NRL_X\"),\n",
    "    (D_A[D_A['chr'] == 'CHROMOSOME_I'], \"CHROMOSOME_I\", \"#16415e\", \"N2_30min_NRL_I\"),\n",
    "    (D_A[D_A['chr'] == 'CHROMOSOME_II'], \"CHROMOSOME_II\", \"#16415e\", \"N2_30min_NRL_II\"),\n",
    "    (D_A[D_A['chr'] == 'CHROMOSOME_III'], \"CHROMOSOME_III\", \"#16415e\", \"N2_30min_NRL_III\"),\n",
    "    (D_A[D_A['chr'] == 'CHROMOSOME_IV'], \"CHROMOSOME_IV\", \"#16415e\", \"N2_30min_NRL_IV\"),\n",
    "    (D_A[D_A['chr'] == 'CHROMOSOME_V'], \"CHROMOSOME_V\", \"#16415e\", \"N2_30min_NRL_V\"),\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "D_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(D_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "D_NRL.columns = cols\n",
    "display(D_NRL)\n",
    "\n",
    "# Assuming D_X, D_A, etc., are your dataframes for chromosomes\n",
    "# You will need to filter D_A for each chromosome I to V\n",
    "chromosome_data=[]\n",
    "chromosome_data = [\n",
    "    (H_X, \"CHROMOSOME_X\", \"#c45746\", \"SDC2-AID-Aux_30min_NRL_X\"),\n",
    "    (H_A[H_A['chr'] == 'CHROMOSOME_I'], \"CHROMOSOME_I\", \"#16415e\", \"AID-SDC2_30min_NRL_I\"),\n",
    "    (H_A[H_A['chr'] == 'CHROMOSOME_II'], \"CHROMOSOME_II\", \"#16415e\", \"AID-SDC2_30min_NRL_II\"),\n",
    "    (H_A[H_A['chr'] == 'CHROMOSOME_III'], \"CHROMOSOME_III\", \"#16415e\", \"AID-SDC2_30min_NRL_III\"),\n",
    "    (H_A[H_A['chr'] == 'CHROMOSOME_IV'], \"CHROMOSOME_IV\", \"#16415e\", \"AID-SDC2_30min_NRL_IV\"),\n",
    "    (H_A[H_A['chr'] == 'CHROMOSOME_V'], \"CHROMOSOME_V\", \"#16415e\", \"AID-SDC2_30min_NRL_V\"),\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "H_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(H_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "H_NRL.columns = cols\n",
    "display(H_NRL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74516c12",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:42:27.861215900Z",
     "start_time": "2023-08-16T01:42:27.857211700Z"
    }
   },
   "outputs": [],
   "source": [
    "### Calculate NRL Length\n",
    "\n",
    "# List of chromosomes to compute the linear regression for\n",
    "chromosomes = ['CHROMOSOME_X', 'CHROMOSOME_I', 'CHROMOSOME_II', 'CHROMOSOME_III', 'CHROMOSOME_IV', 'CHROMOSOME_V']\n",
    "Dequations=[]\n",
    "Hequations=[]\n",
    "# Using D_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(D_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), D_NRL[f'{chrom}-peak-position'].values)\n",
    "    Dslopes = model.coef_[0].round(2)\n",
    "    Dintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Dequations.append(f'{chrom} NRL: {Dslopes}')\n",
    "\n",
    "print(Dequations)\n",
    "\n",
    "# Using H_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(H_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), H_NRL[f'{chrom}-peak-position'].values)\n",
    "    Hslopes = model.coef_[0].round(2)\n",
    "    Hintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Hequations.append(f'{chrom} NRL: {Hslopes}')\n",
    "\n",
    "print(Hequations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db0c0241",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:45:12.418361500Z",
     "start_time": "2023-08-16T01:45:10.627578Z"
    }
   },
   "outputs": [],
   "source": [
    "importlib.reload(nanotools)\n",
    "# Generate plots\n",
    "nanotools.plot_NRL_regression(D_NRL, 'N2 Nucleosome Repeat Length', 'N2_30min',Dequations)\n",
    "nanotools.plot_NRL_regression(H_NRL, 'AID::SDC-2 Nucleosome Repeat Length', 'AID-SDC2_30min',Hequations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d001127",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:03:16.936652600Z",
     "start_time": "2023-08-16T01:02:58.784086200Z"
    }
   },
   "outputs": [],
   "source": [
    "### PLOT NRL DISTRIBUTION BY CHROMOSOME FOR TWO CONDITIONS (N2 and #021)\n",
    "importlib.reload(nanotools)\n",
    "\n",
    "size = (8,12)\n",
    "\n",
    "# Assuming D_X, D_A, etc., are your dataframes for chromosomes\n",
    "# You will need to filter D_A for each chromosome I to V\n",
    "chromosome_data = [\n",
    "    (D_X, \"X\", \"#c45746\", \"N2_30min_NRL_X\"),\n",
    "    (D_A, \"A\", \"#16415e\", \"N2_30min_NRL_A\")\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "D_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(D_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "D_NRL.columns = cols\n",
    "display(D_NRL)\n",
    "\n",
    "# Assuming D_X, D_A, etc., are your dataframes for chromosomes\n",
    "# You will need to filter D_A for each chromosome I to V\n",
    "chromosome_data=[]\n",
    "chromosome_data = [\n",
    "    (H_X, \"X\", \"#c45746\", \"SDC2-AID-Aux_30min_NRL_X\"),\n",
    "    (H_A, \"A\", \"#16415e\", \"AID-SDC2_30min_NRL_A\")\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "H_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(H_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "H_NRL.columns = cols\n",
    "display(H_NRL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7894096a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:41:01.203137200Z",
     "start_time": "2023-08-16T01:40:59.800336700Z"
    }
   },
   "outputs": [],
   "source": [
    "### Calculate NRL Length for X vs A\n",
    "\n",
    "# List of chromosomes to compute the linear regression for\n",
    "chromosomes = ['X', 'A']\n",
    "Dequations=[]\n",
    "Hequations=[]\n",
    "# Using D_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(D_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), D_NRL[f'{chrom}-peak-position'].values)\n",
    "    Dslopes = model.coef_[0].round(2)\n",
    "    Dintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Dequations.append(f'{chrom} NRL: {Dslopes}')\n",
    "\n",
    "print(Dequations)\n",
    "\n",
    "# Using H_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(H_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), H_NRL[f'{chrom}-peak-position'].values)\n",
    "    Hslopes = model.coef_[0].round(2)\n",
    "    Hintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Hequations.append(f'{chrom} NRL: {Hslopes}')\n",
    "\n",
    "print(Hequations)\n",
    "\n",
    "importlib.reload(nanotools)\n",
    "# Generate plots\n",
    "nanotools.plot_NRL_regression(D_NRL, 'N2 Nucleosome Repeat Length', 'N2_30min',Dequations)\n",
    "nanotools.plot_NRL_regression(H_NRL, 'AID::SDC-2 Nucleosome Repeat Length', 'AID-SDC2_30min',Hequations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23b0f2df",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T02:23:21.418599200Z",
     "start_time": "2023-08-16T02:23:19.385805200Z"
    }
   },
   "outputs": [],
   "source": [
    "importlib.reload(nanotools)\n",
    "\n",
    "### Dataframes for TSS specific analysis\n",
    "region_cutoff = 3000\n",
    "D_X_tss = nanotools.filter_nucs_by_features(D_X, bed_tss, region_cutoff)\n",
    "D_A_tss = nanotools.filter_nucs_by_features(D_A, bed_tss, region_cutoff)\n",
    "H_X_tss = nanotools.filter_nucs_by_features(H_X, bed_tss, region_cutoff)\n",
    "H_A_tss = nanotools.filter_nucs_by_features(H_A, bed_tss, region_cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3958db6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T02:28:46.570454100Z",
     "start_time": "2023-08-16T02:28:41.584408Z"
    }
   },
   "outputs": [],
   "source": [
    "### PLOT NRL DISTRIBUTION BY CHROMOSOME FOR TWO CONDITIONS (N2 and #021)\n",
    "importlib.reload(nanotools)\n",
    "\n",
    "size = (8,12)\n",
    "\n",
    "# Use the _tss filtered dataframes\n",
    "chromosome_data = [\n",
    "    (D_X_tss, \"X\", \"#c45746\", \"N2_30min_NRL_X\"),\n",
    "    (D_A_tss, \"A\", \"#16415e\", \"N2_30min_NRL_A\")\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "D_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(D_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "D_NRL.columns = cols\n",
    "display(D_NRL)\n",
    "\n",
    "# Use the _tss filtered dataframes\n",
    "chromosome_data = [\n",
    "    (H_X_tss, \"X\", \"#c45746\", \"SDC2-AID-Aux_30min_NRL_X\"),\n",
    "    (H_A_tss, \"A\", \"#16415e\", \"AID-SDC2_30min_NRL_A\")\n",
    "]\n",
    "\n",
    "peaks_list = []\n",
    "for data, chromosome_name, color, output_prefix in chromosome_data:\n",
    "    peaks = nanotools.plot_NRL_dist(data, chromosome_name, color, output_prefix)\n",
    "    peaks_list.append(peaks)\n",
    "\n",
    "# Concatenate the peaks\n",
    "H_NRL = pd.concat(peaks_list, axis=1)\n",
    "\n",
    "# Iterate through columns and rename 'x' columns\n",
    "cols = list(H_NRL.columns)\n",
    "for i, col in enumerate(cols):\n",
    "    if 'peak-position' in col:\n",
    "        chrom = col.split('-peak-position')[0]\n",
    "        cols[i-1] = f'{chrom}-n-plus'  # rename the 'x' column immediately before the 'peak-position' column\n",
    "\n",
    "H_NRL.columns = cols\n",
    "display(H_NRL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d73b59c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T02:31:26.636175900Z",
     "start_time": "2023-08-16T02:31:25.184443700Z"
    }
   },
   "outputs": [],
   "source": [
    "### Calculate NRL Length for X vs A\n",
    "\n",
    "# List of chromosomes to compute the linear regression for\n",
    "chromosomes = ['X', 'A']\n",
    "Dequations=[]\n",
    "Hequations=[]\n",
    "# Using D_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(D_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), D_NRL[f'{chrom}-peak-position'].values)\n",
    "    Dslopes = model.coef_[0].round(2)\n",
    "    Dintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Dequations.append(f'{chrom} NRL: {Dslopes}')\n",
    "\n",
    "print(Dequations)\n",
    "\n",
    "# Using H_NRL dataframe\n",
    "for chrom in chromosomes:\n",
    "    model.fit(H_NRL[f'{chrom}-n-plus'].values.reshape(-1, 1), H_NRL[f'{chrom}-peak-position'].values)\n",
    "    Hslopes = model.coef_[0].round(2)\n",
    "    Hintercepts = model.intercept_.round(2)\n",
    "    # Append text to Dequations chrom + ' NRL: ' +DslopeA as a single item to list Dequations\n",
    "    Hequations.append(f'{chrom} NRL: {Hslopes}')\n",
    "\n",
    "print(Hequations)\n",
    "\n",
    "importlib.reload(nanotools)\n",
    "# Generate plots\n",
    "nanotools.plot_NRL_regression(D_NRL, 'N2 Nucleosome Repeat Length', 'N2_30min',Dequations)\n",
    "nanotools.plot_NRL_regression(H_NRL, 'AID::SDC-2 Nucleosome Repeat Length', 'AID-SDC2_30min',Hequations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c120cd42",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T01:46:49.542432500Z",
     "start_time": "2023-08-16T01:46:49.527428900Z"
    }
   },
   "outputs": [],
   "source": [
    "# Fit a linear regression model\n",
    "model = LinearRegression()\n",
    "model.fit(D_NRL['peak-number'].to_frame('peak-number'), \n",
    "          D_NRL['X-peak-position'].to_frame('X-peak-position'))\n",
    "DslopeX = model.coef_[0].round(2)\n",
    "DinterceptX = model.intercept_.round(2)\n",
    "DequationX = f'X NRL: {DslopeX}'\n",
    "print(\"DequationX;\",DequationX)\n",
    "\n",
    "model = LinearRegression()\n",
    "model.fit(D_NRL['peak-number'].to_frame('peak-number'), \n",
    "          D_NRL['A-peak-position'].to_frame('A-peak-position'))\n",
    "DslopeA = model.coef_[0].round(1)\n",
    "DinterceptA = model.intercept_.round(1)\n",
    "DequationA = f'A NRL: {DslopeA}'\n",
    "print(\"DequationA;\",DequationA)\n",
    "\n",
    "# Fit a linear regression model\n",
    "model = LinearRegression()\n",
    "model.fit(H_NRL['peak-number'].to_frame('peak-number'), \n",
    "          H_NRL['X-peak-position'].to_frame('X-peak-position'))\n",
    "HslopeX = model.coef_[0].round(1)\n",
    "HinterceptX = model.intercept_.round(1)\n",
    "HequationX = f'X NRL: {HslopeX}'\n",
    "print(\"HequationX;\",HequationX)\n",
    "\n",
    "model = LinearRegression()\n",
    "model.fit(H_NRL['peak-number'].to_frame('peak-number'), \n",
    "          H_NRL['A-peak-position'].to_frame('A-peak-position'))\n",
    "HslopeA = model.coef_[0].round(1)\n",
    "HinterceptA = model.intercept_.round(1)\n",
    "HequationA = f'A NRL: {HslopeA}'\n",
    "print(\"HequationA;\",HequationA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deed9025",
   "metadata": {},
   "outputs": [],
   "source": [
    "### OLD\n",
    "def frequency_table(df):\n",
    "    frequency_df = pd.DataFrame(columns=[\"chr\", \"dist\",\"n-plus\"])\n",
    "\n",
    "    for index, row in df.head(100).iterrows():\n",
    "        # Get the value of the \"smt_pos\" column for this row\n",
    "        chrm = row[\"chr\"]\n",
    "\n",
    "        # Get the value of the \"smt_pos\" column for this row\n",
    "        smt_pos = row[\"smt_pos\"]\n",
    "\n",
    "        # Get the next X rows of the dataframe\n",
    "        next_rows = nucleosome_positions_D[index+1:index+n_dist+1]\n",
    "\n",
    "        n_counter = 0\n",
    "        # Subtract the \"smt_pos\" value from each of the next 10 rows\n",
    "        for i, r in next_rows.iterrows():\n",
    "            # If n'th row's chromosome, and current row's chromosome match:\n",
    "            if r[\"chr\"] == chrm:\n",
    "                n_counter += 1\n",
    "                # calculate nlr\n",
    "                difference = int((r[\"smt_pos\"] - smt_pos)/n_counter)\n",
    "                # append to frequency table\n",
    "                frequency_df = frequency_df.append({\"chr\": chrm, \"dist\": difference, \"n-plus\": n_counter}, ignore_index=True)\n",
    "\n",
    "    # Sort the results_df dataframe by \"smt_pos\" and \"difference\"\n",
    "    #frequency_df = frequency_df.sort_values(by=[\"chr\", \"dist\",\"n-plus\"])\n",
    "\n",
    "    display(frequency_df.tail(20))\n",
    "    return frequency_df\n",
    "\n",
    "\n",
    "###\n",
    "# Fit a linear regression model\n",
    "model = LinearRegression()\n",
    "model.fit(data_D[.loc[data_D['chr-type']=='X','n-plus']].to_frame('n-plus'), \n",
    "          data_D.loc[data_D['chr-type']=='X','dist'].to_frame('dist'))\n",
    "DslopeX = model.coef_[0].round(2)\n",
    "DinterceptX = model.intercept_.round(2)\n",
    "DequationX = f'X: y = {DslopeX}x + {DinterceptX}'\n",
    "print(\"DequationX;\",DequationX)\n",
    "\n",
    "model = LinearRegression()\n",
    "model.fit(data_D.loc[data_D['chr-type']=='A','n-plus'].to_frame('n-plus'), \n",
    "          data_D.loc[data_D['chr-type']=='A','dist'].to_frame('dist'))\n",
    "DslopeA = model.coef_[0].round(1)\n",
    "DinterceptA = model.intercept_.round(1)\n",
    "DequationA = f'A: y = {DslopeA}x + {DinterceptA}'\n",
    "print(\"DequationA;\",DequationA)\n",
    "\n",
    "# Fit a linear regression model\n",
    "model = LinearRegression()\n",
    "model.fit(data_H.loc[data_H['chr-type']=='X','n-plus'].to_frame('n-plus'), \n",
    "          data_H.loc[data_H['chr-type']=='X','dist'].to_frame('dist'))\n",
    "HslopeX = model.coef_[0].round(1)\n",
    "HinterceptX = model.intercept_.round(1)\n",
    "HequationX = f'X: y = {HslopeX}x + {HinterceptX}'\n",
    "print(\"HequationX;\",HequationX)\n",
    "\n",
    "model = LinearRegression()\n",
    "model.fit(data_H.loc[data_H['chr-type']=='A','n-plus'].to_frame('n-plus'), \n",
    "          data_H.loc[data_H['chr-type']=='A','dist'].to_frame('dist'))\n",
    "HslopeA = model.coef_[0].round(1)\n",
    "HinterceptA = model.intercept_.round(1)\n",
    "HequationA = f'A: y = {HslopeA}x + {HinterceptA}'\n",
    "print(\"HequationA;\",HequationA)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
