{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:46:06.158325100Z",
     "start_time": "2023-11-11T02:46:03.497832200Z"
    },
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-2.25.2.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import importlib\n",
    "import nanotools\n",
    "importlib.reload(nanotools) # reload nanotools module\n",
    "#import random\n",
    "#random.seed(10)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.express as px # Used for plotting\n",
    "import plotly.graph_objects as go # Used for plotting\n",
    "from plotly.subplots import make_subplots # Used for plotting\n",
    "from sklearn import cluster # Used for clustering\n",
    "import multiprocessing\n",
    "from multiprocessing import Pool, cpu_count # used for parallel processing\n",
    "from io import StringIO\n",
    "import subprocess\n",
    "import os\n",
    "import plotly.io as pio\n",
    "import plotly\n",
    "import matplotlib.pyplot as plt # Use for plotting m6A frac and coverage plot\n",
    "from matplotlib import cm # Use for plotting m6A frac and coverage plot\n",
    "import sqlite3\n",
    "import tqdm\n",
    "import pysam\n",
    "\n",
    "#import dask.dataframe as dd\n",
    "\n",
    "# set renderer to vscode\n",
    "plotly.offline.init_notebook_mode(connected=True)\n",
    "pio.renderers.default = 'plotly_mimetype+notebook'\n",
    "pd.options.display.max_rows = None\n",
    "pd.set_option('display.max_columns', None)\n",
    "# display count_df with no limits on rows\n",
    "pd.set_option('display.max_rows', None)\n",
    "# left align print\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:46:06.158325100Z",
     "start_time": "2023-11-11T02:46:06.157324700Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Bed file configurations:\n",
    "sample_source = \"chr_type\" # \"chr_type\" or \"type\" or \"chromosome\"\n",
    "sampleName = [\"X\",\"Autosome\"] # \"TES_q1\" \"strong_rex\" \"weak_rex\" \"type\", \"X\", \"Autosome\"; Must be same number of unique values in selected bed rows.\n",
    "chr_type_selected = [\"X\",\"Autosome\"] # 'X' or \"Autosome\"\n",
    "type_selected = [\"TSS_q4\",\"TSS_q3\",\"TSS_q2\",\"TSS_q1\"] #her-1_TSS\",\"fem-1_TSS\",\"fem-2_TSS\",\"fem-3_TSS\",\"sex-1_TSS\",\n",
    "# her-1_TSS/TES/FULL | TES_q1-4 | #TSS_q1-4 | strong/weak rex | whole_chr | 200kb_region | 50kb_region | center_DPY27_chip_albretton | gene_q1-q4 | MEX_motif | center_SDC2_chip_albretton | center_SDC3_chip_albretton |\n",
    "max_regions = 0 # max regions to consider; 0 = full set;\n",
    "chromosome_selected = [\"CHROMOSOME_I\", \"CHROMOSOME_II\", \"CHROMOSOME_III\", \"CHROMOSOME_IV\",\"CHROMOSOME_V\",\"CHROMOSOME_X\"]\n",
    "strand_selected = [\"+\",\"-\"] #+ and/or -\n",
    "select_opp_strand = True #If you want to select both + and - strands for all regions set to True\n",
    "down_sample_autosome = True # If you want to downsample autosome genes to match number of X genes set to True\n",
    "if chr_type_selected == [\"X\"]:\n",
    "    down_sample_autosome = False\n",
    "bed_file = \"/Data1/reference/tss_tes_rex_combined_v11_WS235.bed\"\n",
    "bed_window = 1000 # +/- around bed elements.\n",
    "mods = \"a\" # {A,CG,A+CG}\n",
    "if sample_source == \"chr_type\":\n",
    "    selection = chr_type_selected\n",
    "if sample_source == \"type\":\n",
    "    selection = type_selected\n",
    "if sample_source == \"chromosome\":\n",
    "    selection = chromosome_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:46:06.158325100Z",
     "start_time": "2023-11-11T02:46:06.158325100Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nif sample_source == \"chr_type\":\\n    selection = chr_type_selected\\nif sample_source == \"type\":\\n    selection = type_selected\\nif sample_source == \"chromosome\":\\n    selection = chromosome_selected\\nif sample_source == \"condition\":\\n    selection = analysis_cond'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### BAM Configurations\n",
    "m6A_thresh = 181 #default is 129 = 50%; 181=70%; 194=75%; 207 = 80%; 232 = 90%\n",
    "mC_thresh = 129 #default is 129\n",
    "\n",
    "# modkit is used for aggregating methylation data from .bam files\n",
    "# https://nanoporetech.github.io/modkit/quick_start.html\n",
    "modkit_path = \"/Data1/software/modkit/modkit\"\n",
    "\n",
    "### IMPORT BAM FILES AND METADATA FROM CSV FILE\n",
    "input_metadata = pd.read_csv(\"/Data1/seq_data/combined_fiber_dimelo_8_30_23/input_metadata_8_30_23.txt\", sep=\"\\t\", header=0)\n",
    "# Options: N2_fiber; SDC2_degron_fiber; SDC2_degron_bg; N2_bg; N2-DPY27_dimelo_pAHia5\n",
    "analysis_cond = [\"N2_fiber\",\"SDC2_degron_fiber\"]#,\"N2_young\"]\n",
    "# Set bam_files equal to list of items in column bam_files where conditions == N2_fiber\n",
    "bam_files = input_metadata[input_metadata[\"conditions\"].isin(analysis_cond)][\"bam_files\"].tolist()\n",
    "conditions = input_metadata[input_metadata[\"conditions\"].isin(analysis_cond)][\"conditions\"].tolist()\n",
    "exp_ids = input_metadata[input_metadata[\"conditions\"].isin(analysis_cond)][\"exp_id_date\"].tolist()\n",
    "bam_fracs = len(bam_files)*[1] # For full .bam set to = 1\n",
    "sample_indices = list(range(len(bam_files)))\n",
    "output_stem = \"/Data1/seq_data/combined_fiber_dimelo_8_30_23/\"\n",
    "thresh_list=len(bam_files)*[m6A_thresh/258] # For R10 flow cells use 0.5; for R9 flow cells use 0.9\n",
    "file_prefix = \"aug_27_\"\n",
    "\"\"\"\n",
    "if sample_source == \"chr_type\":\n",
    "    selection = chr_type_selected\n",
    "if sample_source == \"type\":\n",
    "    selection = type_selected\n",
    "if sample_source == \"chromosome\":\n",
    "    selection = chromosome_selected\n",
    "if sample_source == \"condition\":\n",
    "    selection = analysis_cond\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:46:06.999563500Z",
     "start_time": "2023-11-11T02:46:06.298029200Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| temp_bed | first 5, random 5 and last 5 out of total 2289 rows.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>chr-type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>2623</td>\n",
       "      <td>4623</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>8303</td>\n",
       "      <td>10303</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>23953</td>\n",
       "      <td>25953</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>33439</td>\n",
       "      <td>35439</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>38220</td>\n",
       "      <td>40220</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q1</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>13411566</td>\n",
       "      <td>13413566</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>8783527</td>\n",
       "      <td>8785527</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>5465582</td>\n",
       "      <td>5467582</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>12467066</td>\n",
       "      <td>12469066</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>4189868</td>\n",
       "      <td>4191868</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17670018</td>\n",
       "      <td>17672018</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17681283</td>\n",
       "      <td>17683283</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17694221</td>\n",
       "      <td>17696221</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713445</td>\n",
       "      <td>17715445</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713954</td>\n",
       "      <td>17715954</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome     start       end strand    type chr-type\n",
       "0   CHROMOSOME_X      2623      4623      +  TSS_q2        X\n",
       "1   CHROMOSOME_X      8303     10303      -  TSS_q3        X\n",
       "2   CHROMOSOME_X     23953     25953      -  TSS_q2        X\n",
       "3   CHROMOSOME_X     33439     35439      -  TSS_q2        X\n",
       "4   CHROMOSOME_X     38220     40220      +  TSS_q1        X\n",
       "5   CHROMOSOME_X  13411566  13413566      +  TSS_q2        X\n",
       "6   CHROMOSOME_X   8783527   8785527      +  TSS_q2        X\n",
       "7   CHROMOSOME_X   5465582   5467582      +  TSS_q4        X\n",
       "8   CHROMOSOME_X  12467066  12469066      +  TSS_q3        X\n",
       "9   CHROMOSOME_X   4189868   4191868      +  TSS_q4        X\n",
       "10  CHROMOSOME_X  17670018  17672018      +  TSS_q3        X\n",
       "11  CHROMOSOME_X  17681283  17683283      -  TSS_q2        X\n",
       "12  CHROMOSOME_X  17694221  17696221      +  TSS_q2        X\n",
       "13  CHROMOSOME_X  17713445  17715445      +  TSS_q3        X\n",
       "14  CHROMOSOME_X  17713954  17715954      +  TSS_q4        X"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Data1/reference/X.bed has been compressed successfully to /Data1/reference/X.bed.gz\n",
      "Index created successfully for /Data1/reference/X.bed.gz.tbi\n",
      "| temp_bed | first 5, random 5 and last 5 out of total 12990 rows.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>chr-type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>9232</td>\n",
       "      <td>11232</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>9413</td>\n",
       "      <td>11413</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>25781</td>\n",
       "      <td>27781</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>34350</td>\n",
       "      <td>36350</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>42733</td>\n",
       "      <td>44733</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>10787879</td>\n",
       "      <td>10789879</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>9546086</td>\n",
       "      <td>9548086</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q4</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHROMOSOME_IV</td>\n",
       "      <td>17489489</td>\n",
       "      <td>17491489</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>18461797</td>\n",
       "      <td>18463797</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHROMOSOME_II</td>\n",
       "      <td>127706</td>\n",
       "      <td>129706</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q1</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>20874273</td>\n",
       "      <td>20876273</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>20897860</td>\n",
       "      <td>20899860</td>\n",
       "      <td>-</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>20898396</td>\n",
       "      <td>20900396</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q2</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>20904886</td>\n",
       "      <td>20906886</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>20914328</td>\n",
       "      <td>20916328</td>\n",
       "      <td>+</td>\n",
       "      <td>TSS_q3</td>\n",
       "      <td>Autosome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome     start       end strand    type  chr-type\n",
       "0    CHROMOSOME_I      9232     11232      -  TSS_q3  Autosome\n",
       "1    CHROMOSOME_I      9413     11413      +  TSS_q3  Autosome\n",
       "2    CHROMOSOME_I     25781     27781      -  TSS_q3  Autosome\n",
       "3    CHROMOSOME_I     34350     36350      -  TSS_q4  Autosome\n",
       "4    CHROMOSOME_I     42733     44733      +  TSS_q2  Autosome\n",
       "5    CHROMOSOME_I  10787879  10789879      +  TSS_q4  Autosome\n",
       "6    CHROMOSOME_V   9546086   9548086      +  TSS_q4  Autosome\n",
       "7   CHROMOSOME_IV  17489489  17491489      -  TSS_q3  Autosome\n",
       "8    CHROMOSOME_V  18461797  18463797      +  TSS_q2  Autosome\n",
       "9   CHROMOSOME_II    127706    129706      -  TSS_q1  Autosome\n",
       "10   CHROMOSOME_V  20874273  20876273      +  TSS_q3  Autosome\n",
       "11   CHROMOSOME_V  20897860  20899860      -  TSS_q2  Autosome\n",
       "12   CHROMOSOME_V  20898396  20900396      +  TSS_q2  Autosome\n",
       "13   CHROMOSOME_V  20904886  20906886      +  TSS_q3  Autosome\n",
       "14   CHROMOSOME_V  20914328  20916328      +  TSS_q3  Autosome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Data1/reference/Autosome.bed has been compressed successfully to /Data1/reference/Autosome.bed.gz\n",
      "Index created successfully for /Data1/reference/Autosome.bed.gz.tbi\n",
      "Saved the following bedfiles: ['/Data1/reference/X.bed.gz', '/Data1/reference/Autosome.bed.gz']\n",
      "| modkit_bed_df | first 5, random 5 and last 5 out of total 9156 rows.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>42733</td>\n",
       "      <td>44733</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>42733</td>\n",
       "      <td>44733</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>63021</td>\n",
       "      <td>65021</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>63021</td>\n",
       "      <td>65021</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHROMOSOME_I</td>\n",
       "      <td>90380</td>\n",
       "      <td>92380</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>4523221</td>\n",
       "      <td>4525221</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHROMOSOME_V</td>\n",
       "      <td>127009</td>\n",
       "      <td>129009</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>7891389</td>\n",
       "      <td>7893389</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>14089380</td>\n",
       "      <td>14091380</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>10122404</td>\n",
       "      <td>10124404</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17694221</td>\n",
       "      <td>17696221</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713445</td>\n",
       "      <td>17715445</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713445</td>\n",
       "      <td>17715445</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713954</td>\n",
       "      <td>17715954</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHROMOSOME_X</td>\n",
       "      <td>17713954</td>\n",
       "      <td>17715954</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0         1         2  3  4  5\n",
       "0   CHROMOSOME_I     42733     44733  .  .  +\n",
       "1   CHROMOSOME_I     42733     44733  .  .  -\n",
       "2   CHROMOSOME_I     63021     65021  .  .  -\n",
       "3   CHROMOSOME_I     63021     65021  .  .  +\n",
       "4   CHROMOSOME_I     90380     92380  .  .  +\n",
       "5   CHROMOSOME_X   4523221   4525221  .  .  +\n",
       "6   CHROMOSOME_V    127009    129009  .  .  +\n",
       "7   CHROMOSOME_X   7891389   7893389  .  .  -\n",
       "8   CHROMOSOME_X  14089380  14091380  .  .  -\n",
       "9   CHROMOSOME_X  10122404  10124404  .  .  +\n",
       "10  CHROMOSOME_X  17694221  17696221  .  .  -\n",
       "11  CHROMOSOME_X  17713445  17715445  .  .  +\n",
       "12  CHROMOSOME_X  17713445  17715445  .  .  -\n",
       "13  CHROMOSOME_X  17713954  17715954  .  .  +\n",
       "14  CHROMOSOME_X  17713954  17715954  .  .  -"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program finished!\n",
      "['/Data1/seq_data/TubeY9B_N2_fiberseq_timec_mpx_3_21_2023/demux/mod_mappings_barcode_08_Ya_ambig.bam', '/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/basecalls/m6A/mod_mappings.sorted_Ya_ambig.bam', '/Data1/seq_data/TubeAD1_N2_fiberseq_6_13_23/mod_basecalls/mod_mappings.sorted.bam', '/Data1/seq_data/TubeAB_FiberSeq_TimeC_N2_021Aux_4_10_23/basecalls/m6A_full/demux/mod_mappings_barcode_05_Ya_ambig.bam', '/Data1/seq_data/TubeAB_FiberSeq_TimeC_N2_021Aux_4_10_23/basecalls/m6A_full/demux/mod_mappings_barcode_04_Ya_ambig.bam', '/Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/8_20_23_run/pod5_pass/barcode07/basecalls/barcode07.mod_mappings.sorted.bam', '/Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/8_20_23_run/pod5_pass/barcode08/basecalls/barcode08.mod_mappings.sorted.bam', '/Data1/seq_data/TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19/basecalls/m6A/mod_mappings.sorted_Ya_ambig.bam', '/Data1/seq_data/TubeAB_FiberSeq_TimeC_N2_021Aux_4_10_23/basecalls/m6A_full/demux/mod_mappings_barcode_10_Ya_ambig.bam', '/Data1/seq_data/TubeAB_FiberSeq_TimeC_N2_021Aux_4_10_23/basecalls/m6A_full/demux/mod_mappings_barcode_09_Ya_ambig.bam', '/Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/8_20_23_run/pod5_pass/barcode09/basecalls/barcode09.mod_mappings.sorted.bam']\n"
     ]
    }
   ],
   "source": [
    "# Filter input bed_file based on input parameters (e.g. chromosome, type, strand, etc.)\n",
    "# Function saves a new filtered bed file to the same folder as the original bed file\n",
    "# called temp_do_not_use_\"type\".bed\n",
    "importlib.reload(nanotools)\n",
    "new_bed_files=nanotools.filter_bed_file(\n",
    "    bed_file,\n",
    "    sample_source,\n",
    "    selection,\n",
    "    chromosome_selected,\n",
    "    chr_type_selected,\n",
    "    type_selected,\n",
    "    strand_selected,\n",
    "    max_regions,\n",
    "    bed_window\n",
    ")\n",
    "\n",
    "modkit_bed_name = \"modkit_temp.bed\"\n",
    "modkit_bed_df = nanotools.generate_modkit_bed(new_bed_files, down_sample_autosome, select_opp_strand,modkit_bed_name)\n",
    "nanotools.display_sample_rows(modkit_bed_df, 5)\n",
    "\n",
    "# Subsample bam based on bam_frac, used to accelerate testing\n",
    "# if bam_frac = 1 will use original bam files, otherwise will save new subsampled bam files to output_stem.\n",
    "args_list = [(bam_file, condition, bam_frac, sample_index, output_stem) for bam_file, condition, bam_frac, sample_index in zip(bam_files,conditions,bam_fracs,sample_indices)]\n",
    "new_bam_files=[]\n",
    "if __name__ == \"__main__\":\n",
    "    with Pool() as pool: #processes=1\n",
    "        new_bam_files = pool.starmap(nanotools.subsample_bam, args_list)\n",
    "    print(\"Program finished!\")\n",
    "\n",
    "print(new_bam_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:47:09.717421700Z",
     "start_time": "2023-11-11T02:46:08.979153900Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_4_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_3_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedStarting on: /Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/8_20_23_run/pod5_pass/barcode08/basecalls/barcode08.mod_mappings.sorted.bamSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_2_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_5_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_1_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-N2_fiber_0.7_0_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-SDC2_degron_fiber_0.7_7_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-SDC2_degron_fiber_0.7_9_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedSkipping: /Data1/seq_data/combined_fiber_dimelo_8_30_23/modkit-extract-SDC2_degron_fiber_0.7_8_1TSS_q4-TSS_q3-TSS_q2-TSS_q1_CE_I-C_II-CIII-C_IV-CE_V-CE_X_.bedStarting on: /Data1/seq_data/AH_N2_SDC2aid_AuxRem_fiberseq_8_19_23/8_20_23_run/pod5_pass/barcode09/basecalls/barcode09.mod_mappings.sorted.bam\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> > specifying include-only BED outputs only mapped sitesspecifying include-only BED outputs only mapped sites\n",
      "\n",
      "> > parsing BED at parsing BED at modkit_temp.bedmodkit_temp.bed\n",
      "\n",
      "> processed 9156 BED lines\n",
      "> processed 9156 BED lines\n",
      "> found BAM index, processing reads in 100000 base pair chunks\n",
      "> found BAM index, processing reads in 100000 base pair chunks\n",
      "> processed 65490 reads, 1274670 rows, skipped ~76354 reads, failed ~0 reads\n",
      "Process ForkPoolWorker-28:\n",
      "Process ForkPoolWorker-45:\n",
      "Process ForkPoolWorker-44:\n",
      "Process ForkPoolWorker-25:\n",
      "Process ForkPoolWorker-47:\n",
      "Process ForkPoolWorker-42:\n",
      "Process ForkPoolWorker-46:\n",
      "Process ForkPoolWorker-26:\n",
      "Process ForkPoolWorker-37:\n",
      "Process ForkPoolWorker-43:\n",
      "Process ForkPoolWorker-40:\n",
      "Process ForkPoolWorker-48:\n",
      "Process ForkPoolWorker-41:\n",
      "Process ForkPoolWorker-39:\n",
      "Process ForkPoolWorker-38:\n",
      "Process ForkPoolWorker-36:\n",
      "Process ForkPoolWorker-30:\n",
      "Process ForkPoolWorker-33:\n",
      "Process ForkPoolWorker-32:\n",
      "Process ForkPoolWorker-29:\n",
      "Process ForkPoolWorker-31:\n",
      "Process ForkPoolWorker-27:\n",
      "Process ForkPoolWorker-34:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 352, in get\n",
      "    res = self._reader.recv_bytes()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 110, in worker\n",
      "    task = get()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/connection.py\", line 216, in recv_bytes\n",
      "    buf = self._recv_bytes(maxlength)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/connection.py\", line 407, in _recv_bytes\n",
      "    buf = self._recv(4)\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/queues.py\", line 351, in get\n",
      "    with self._rlock:\n",
      "Process ForkPoolWorker-35:\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/connection.py\", line 379, in _recv\n",
      "    chunk = read(handle, remaining)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 297, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/process.py\", line 99, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 121, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\", line 44, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/tmp/ipykernel_10834/1766281913.py\", line 44, in modkit_extract\n",
      "    subprocess.run(command, text=True)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/subprocess.py\", line 490, in run\n",
      "    stdout, stderr = process.communicate(input, timeout=timeout)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/subprocess.py\", line 956, in communicate\n",
      "    self.wait()\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/subprocess.py\", line 1019, in wait\n",
      "    return self._wait(timeout=timeout)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/subprocess.py\", line 1653, in _wait\n",
      "    (pid, sts) = self._try_wait(0)\n",
      "  File \"/home/ubuntu/anaconda3/envs/nanopore_NRL/lib/python3.7/subprocess.py\", line 1611, in _try_wait\n",
      "    (pid, sts) = os.waitpid(self.pid, wait_flags)\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_10834/1766281913.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mPool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m     \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodkit_extract\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36mmap\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    267\u001b[0m         '''\n\u001b[0;32m--> 268\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_map_async\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0miterable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmapstar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchunksize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    269\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    650\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 651\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    652\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mready\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    647\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 648\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_event\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    649\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    551\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 552\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    553\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    295\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 296\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_10834/1766281913.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     59\u001b[0m \u001b[0;31m# Execute commands in parallel\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mPool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m     \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodkit_extract\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"finished with:\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36m__exit__\u001b[0;34m(self, exc_type, exc_val, exc_tb)\u001b[0m\n\u001b[1;32m    621\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    622\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__exit__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_val\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_tb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 623\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mterminate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    624\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    625\u001b[0m \u001b[0;31m#\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36mterminate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    546\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTERMINATE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    547\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_worker_handler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTERMINATE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 548\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_terminate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    549\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    550\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/util.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, wr, _finalizer_registry, sub_debug, getpid)\u001b[0m\n\u001b[1;32m    222\u001b[0m                 sub_debug('finalizer calling %s with args %s and kwargs %s',\n\u001b[1;32m    223\u001b[0m                           self._callback, self._args, self._kwargs)\n\u001b[0;32m--> 224\u001b[0;31m                 \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_callback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    225\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_weakref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_callback\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_args\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m                             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_kwargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_key\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36m_terminate_pool\u001b[0;34m(cls, taskqueue, inqueue, outqueue, pool, worker_handler, task_handler, result_handler, cache)\u001b[0m\n\u001b[1;32m    579\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    580\u001b[0m         \u001b[0mutil\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'helping task handler/workers to finish'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 581\u001b[0;31m         \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_help_stuff_finish\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minqueue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_handler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    582\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    583\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;32mnot\u001b[0m \u001b[0mresult_handler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_alive\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcache\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/nanopore_NRL/lib/python3.7/multiprocessing/pool.py\u001b[0m in \u001b[0;36m_help_stuff_finish\u001b[0;34m(inqueue, task_handler, size)\u001b[0m\n\u001b[1;32m    564\u001b[0m         \u001b[0;31m# task_handler may be blocked trying to put items on inqueue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    565\u001b[0m         \u001b[0mutil\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'removing tasks from inqueue until task handler finished'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 566\u001b[0;31m         \u001b[0minqueue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_rlock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    567\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mtask_handler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_alive\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0minqueue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpoll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    568\u001b[0m             \u001b[0minqueue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "### Extracting per read modifications\n",
    "out_file_names = [output_stem + \"modkit-extract-\" + each_condition +\"_\"+ str(round(each_thresh,2))+\"_\"+str(each_index)+ \"_\"+str(each_bamfrac)+\n",
    "                  # convert the first 3 characters of each element in \"type_selected\" into a single string separated by \"-\"\n",
    "                  \"-\".join([str(x)[0:7] for x in type_selected])+\"_\"+\n",
    "                  # convert first character and the last 3 characters of each element in \"choromosome_selected\" into a single string separated by \"-\"\n",
    "                  \"-\".join([str(x)[0]+str(x)[-3:] for x in chromosome_selected])+\"_\"+\n",
    "                  \".bed\"\n",
    "                  for each_condition,each_thresh,each_index, each_bamfrac in zip(conditions,thresh_list,sample_indices,bam_fracs)]\n",
    "\n",
    "modkit_bed_df = pd.read_csv(modkit_bed_name,sep='\\t',header=None)\n",
    "### Define bed file for modkit\n",
    "\n",
    "# Function to run a single extract command\n",
    "def modkit_extract(args):\n",
    "    each_bam, each_thresh, each_condition, each_index, each_bamfrac,modkit_path, output_stem, modkit_bed_name, bed_window = args\n",
    "\n",
    "    each_output = output_stem + \"modkit-extract-\" + each_condition +\"_\"+ str(round(each_thresh,2))+\"_\"+str(each_index)+ \"_\"+str(each_bamfrac)+ \"-\".join([str(x)[0:7] for x in type_selected])+\"_\"+ \"-\".join([str(x)[0]+str(x)[-3:] for x in chromosome_selected])+\"_\"+ \".bed\"\n",
    "\n",
    "    ### NOTE: Name of pileup file is not based on configurations\n",
    "    ### TODO: Name of output file should be based on configs so that we aren't recomputing pileups withidentical conditions.\n",
    "\n",
    "    # If each_output exsits, skip\n",
    "    if os.path.exists(each_output):\n",
    "        print(f\"Skipping: {each_output}\")\n",
    "        return\n",
    "\n",
    "    print(f\"Starting on: {each_bam}\")\n",
    "    command = [\n",
    "        modkit_path,\n",
    "        \"extract\",\n",
    "        \"--threads\",\n",
    "        \"20\",\n",
    "        \"--force\",\n",
    "        \"--mapped\",\n",
    "        \"--ignore\",\n",
    "        \"m\",\n",
    "        \"--include-bed\",\n",
    "        modkit_bed_name,\n",
    "        \"--log-filepath\",\n",
    "        each_output + each_condition + \"_modkit-extract.log\",\n",
    "        each_bam,\n",
    "        each_output\n",
    "    ]\n",
    "    subprocess.run(command, text=True)\n",
    "\n",
    "    # Create a list of arguments for each task\n",
    "task_args = list(zip(\n",
    "    new_bam_files,\n",
    "    thresh_list,\n",
    "    conditions,\n",
    "    sample_indices,\n",
    "    bam_fracs,\n",
    "    [modkit_path]*len(new_bam_files),\n",
    "    [output_stem]*len(new_bam_files),\n",
    "    [modkit_bed_name]*len(new_bam_files),\n",
    "    [bed_window]*len(new_bam_files)\n",
    "))\n",
    "\n",
    "# Execute commands in parallel\n",
    "with Pool() as pool:\n",
    "    pool.map(modkit_extract, task_args)\n",
    "\n",
    "print(\"finished with:\")\n",
    "print(out_file_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:45:43.235624400Z",
     "start_time": "2023-11-11T02:45:43.220657400Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_10834/1381115496.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0;31m# Initialize comb_readmethyl_df as an empty dataframe\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m \u001b[0mcomb_readmethyl_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;31m# Build combined bed df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "### Add bed and condition details to modkit output for plotting\n",
    "# Using DataFrame merging to achieve the task without explicit loops\n",
    "## Looks up the bed_start and bed_end values for each row in bedmethyl_df\n",
    "def add_bed_columns_no_loops(bedmethyl_df, bed_df):\n",
    "\n",
    "    # Initialize an empty DataFrame to store the merged data\n",
    "    merged_data = pd.DataFrame()\n",
    "    filtered_df = pd.DataFrame()\n",
    "\n",
    "    # drop all rows from bedmethly_df where mod_strand != '+' or '-'\n",
    "    bedmethyl_df = bedmethyl_df[(bedmethyl_df['mod_strand'] == '+') | (bedmethyl_df['mod_strand'] == '-')]\n",
    "    # Get all unique chromosome-strand combinations without headers in bedmethyl_df\n",
    "    unique_chrom_strand_combinations = bedmethyl_df[['chrom']].drop_duplicates().values #, 'mod_strand'\n",
    "    for chrom in unique_chrom_strand_combinations: #, strand\n",
    "        #print(\"Starting on:\", chrom, strand)\n",
    "        bed_subset = bed_df[(bed_df['chrom'] == chrom[0])] #& (bed_df['bed_strand'] == strand)]\n",
    "\n",
    "        if bed_subset.empty:\n",
    "            #print(\"No bed entries for {}\".format(chrom[0]))\n",
    "            # then skip this chromosome-strand combination\n",
    "            continue\n",
    "        #display(bed_subset.head(10))\n",
    "        # Subset data for the current chromosome\n",
    "        bedmethyl_subset = bedmethyl_df[(bedmethyl_df['chrom'] == chrom[0])]\n",
    "        #print(\"bedmethyl_subset:\",bedmethyl_subset.head(10))\n",
    "        #display(bedmethyl_subset.head(10))\n",
    "        # Explicitly cast to numeric data type\n",
    "        bedmethyl_subset['ref_position'] = bedmethyl_subset['ref_position'].astype('int64')\n",
    "        bed_subset['bed_start'] = bed_subset['bed_start'].astype('int64')\n",
    "\n",
    "        # Perform the merge for the current subset\n",
    "        merged_subset = pd.merge_asof(\n",
    "            bedmethyl_subset,\n",
    "            bed_subset[['bed_start', 'bed_end', 'bed_strand','chr_type','type']],  # Exclude 'chrom' from right DF\n",
    "            left_on='ref_position',\n",
    "            right_on='bed_start',\n",
    "            direction='nearest'\n",
    "        )\n",
    "\n",
    "        # Append the merged data to the overall result\n",
    "        merged_data = merged_data.append(merged_subset)\n",
    "\n",
    "    #print(\"bedmethyl_df:\")\n",
    "    #display(merged_data.head(10))\n",
    "    # Filter out rows where ref_position is not within bed_start and bed_end\n",
    "    filtered_df = merged_data.loc[\n",
    "        (merged_data['ref_position'] >= merged_data['bed_start']) &\n",
    "        (merged_data['ref_position'] <= merged_data['bed_end'])\n",
    "    ]\n",
    "    # Filter out rows where bed_start is not a number\n",
    "    filtered_df = filtered_df[filtered_df['bed_start'].notna()]\n",
    "    filtered_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    # Print number rows in filtered_df\n",
    "    print(\"Found {} rows in filtered_df\".format(len(filtered_df)))\n",
    "\n",
    "    #display(filtered_df)\n",
    "    return filtered_df\n",
    "\n",
    "# Initialize comb_readmethyl_df as an empty dataframe\n",
    "comb_readmethyl_df = pd.DataFrame()\n",
    "\n",
    "# Build combined bed df\n",
    "combined_bed_df=pd.DataFrame()\n",
    "for each_bed in new_bed_files:\n",
    "    bed_path = each_bed[:-3] # remove .gz\n",
    "    print(\"Starting on:\",bed_path)\n",
    "    # read bedpath and append to bed_df\n",
    "    combined_bed_df = combined_bed_df.append(pd.read_csv(bed_path, sep=\"\\t\", header=None))#skiprows=1))\n",
    "combined_bed_df.columns = ['chrom','bed_start','bed_end','bed_strand','type','chr_type']\n",
    "\n",
    "combined_bed_df.sort_values(['chrom','bed_start'], inplace=True)\n",
    "combined_bed_df.reset_index(drop=True, inplace=True)\n",
    "display(combined_bed_df.head(10))\n",
    "\n",
    "# Initialize comb_bedmethyl_plot_df\n",
    "comb_bedmethyl_plot_df = pd.DataFrame()\n",
    "\n",
    "# Create combined plotting dataframe\n",
    "for each_output,each_condition in zip(out_file_names,conditions):\n",
    "    print(\"Starting on:\",each_output)\n",
    "    # Define bed methyl columns and import bedmethyl file\n",
    "    bedmethyl_df = pd.DataFrame()\n",
    "    bedmethyl_cols = ['read_id',\n",
    "    'forward_read_position',\n",
    "    'ref_position',\n",
    "    'chrom',\n",
    "    'mod_strand',\n",
    "    'ref_strand',\n",
    "    'ref_mod_strand',\n",
    "    'fw_soft_clipped_start',\n",
    "    'fw_soft_clipped_end',\n",
    "    'read_length',\n",
    "    'mod_qual',\n",
    "    'mod_code',\n",
    "    'base_qual',\n",
    "    'ref_kmer',\n",
    "    'query_kmer',\n",
    "    'canonical_base',\n",
    "    'modified_primary_base',\n",
    "    'inferred']\n",
    "    bedmethyl_df=pd.read_csv(each_output, sep=\"\\t\", header=None, names=bedmethyl_cols,skiprows=1)\n",
    "\n",
    "    # sort bedmethyl_df by chrom and start_position, required for merging in bed matching.\n",
    "    bedmethyl_df.sort_values(['chrom','ref_position'], inplace=True)\n",
    "    bedmethyl_df.reset_index(drop=True, inplace=True)\n",
    "    #display(bedmethyl_df.head(10))\n",
    "\n",
    "    # Adding new columns to bedmethyl_df using condition matching\n",
    "    bedmethyl_df = add_bed_columns_no_loops(bedmethyl_df.copy(), combined_bed_df)\n",
    "    # Add rel_start and rel_end columns equal to start-bed_start and end-bed_start\n",
    "    bedmethyl_df['rel_pos'] = bedmethyl_df['ref_position'] - bedmethyl_df['bed_start'] -bed_window +1\n",
    "\n",
    "    # add condition column\n",
    "    bedmethyl_df['condition'] = each_condition\n",
    "\n",
    "    # if comb_bedmethyl_plot_df is null, set it equal to bedmethyl_plot\n",
    "    if comb_bedmethyl_plot_df.empty:\n",
    "        print(\"comb_bedmethyl_plot_df is empty, setting it equal to bedmethyl_plot...\")\n",
    "        comb_bedmethyl_plot_df = bedmethyl_df.copy()\n",
    "    # else append bedmethyl_plot to comb_bedmethyl_plot_df\n",
    "    else:\n",
    "        print(\"comb_bedmethyl_plot_df is not empty, appending bedmethyl_plot...\")\n",
    "        comb_bedmethyl_plot_df = comb_bedmethyl_plot_df.append(bedmethyl_df)\n",
    "\n",
    "if any(x in type_selected[0] for x in (\"TSS\", \"TES\", \"MEX\")):\n",
    "    print(\"Strand orientation sensitive type selected, multiplying rel_start by -1 for '-' strand genes...\")\n",
    "    mask = comb_bedmethyl_plot_df['bed_strand'] == '-'\n",
    "    comb_bedmethyl_plot_df.loc[mask, 'rel_pos'] *= -1\n",
    "    #comb_bedmethyl_plot_df.sort_values(['chrom','rel_pos'], inplace=True)\n",
    "\n",
    "# Set entire mod_qual column equal to 1 base on m6A thresh\n",
    "threshold = m6A_thresh / 255\n",
    "# convert comb_bedmethly_plot_df_final['mod_qual'] to a float\n",
    "comb_bedmethyl_plot_df['mod_qual'] = comb_bedmethyl_plot_df['mod_qual'].astype(float)\n",
    "comb_bedmethyl_plot_df['mod_qual_bin'] = np.where(comb_bedmethyl_plot_df['mod_qual'] > threshold, 1, 0)\n",
    "\n",
    "#add read_start and read_end columns to comb_bedmethyl_plot_df_final based on min and max ref_position for each read_id\n",
    "print(\"Calculating read start and end...\")\n",
    "# Reduce the DataFrame size by selecting only the columns you need\n",
    "small_df = comb_bedmethyl_plot_df[['read_id', 'rel_pos']]\n",
    "# Perform the grouping and aggregation in one step\n",
    "grouped_df = small_df.groupby('read_id')['rel_pos'].agg(['min', 'max']).reset_index()\n",
    "# Rename the columns\n",
    "grouped_df.rename(columns={'min': 'rel_read_start', 'max': 'rel_read_end'}, inplace=True)\n",
    "# Merge the aggregated results back to the original DataFrame\n",
    "comb_bedmethyl_plot_df = pd.merge(comb_bedmethyl_plot_df, grouped_df, on='read_id', how='left')\n",
    "# delete small_df and grouped_df\n",
    "del small_df\n",
    "del grouped_df\n",
    "\n",
    "#comb_bedmethyl_plot_df.sort_values(['chrom','rel_pos'], inplace=True)\n",
    "comb_bedmethyl_plot_df = comb_bedmethyl_plot_df[comb_bedmethyl_plot_df['bed_start'].notna()]\n",
    "comb_bedmethyl_plot_df.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# print count of unique read_ids for each condition\n",
    "print(\"All unique read_ids for each condition:\")\n",
    "print(comb_bedmethyl_plot_df.groupby(['condition'])['read_id'].nunique())\n",
    "\n",
    "#display(comb_bedmethyl_plot_df.head(10))\n",
    "# drop all rows where read_length is less than 1000\n",
    "comb_bedmethyl_plot_df = comb_bedmethyl_plot_df[comb_bedmethyl_plot_df['read_length'] >= 1000]\n",
    "\n",
    "\n",
    "# print count of unique read_ids for each condition\n",
    "print(\"All unique read_ids for each condition > read_len of 1000:\")\n",
    "print(comb_bedmethyl_plot_df.groupby(['condition'])['read_id'].nunique())\n",
    "\n",
    "# Define SQLite database file name\n",
    "#db_fn = \"temp_files/\" + \"plot_db_\" + \"-\".join([str(x)[:6] for x in type_selected]) + \"_\" + str(round(thresh_list[0],2)) + \"_\" + str(bam_fracs[0]) + \".db\"\n",
    "# Create a SQLite database connection\n",
    "#conn = sqlite3.connect(db_fn)\n",
    "# Save the DataFrame to SQLite database\n",
    "#comb_bedmethyl_plot_df.to_sql('bedmethyl_plot', conn, if_exists='replace', index=False)\n",
    "# Close the connection\n",
    "#conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-11T02:45:27.139127500Z",
     "start_time": "2023-11-11T02:45:27.072517400Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'type_selected' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_10834/2604234629.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Similar to other dataframes, define fn based on configurations. If it doesn't exist, create it, otherwise import it.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mplot_df_fn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"temp_files/\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"plot_df_\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"-\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m12\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtype_selected\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mthresh_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_fracs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0;34m\".csv\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mcomb_bedmethyl_plot_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mplot_df_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mplot_df_fn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"plot_df_fn exists, importing...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'type_selected' is not defined"
     ]
    }
   ],
   "source": [
    "# Similar to other dataframes, define fn based on configurations. If it doesn't exist, create it, otherwise import it.\n",
    "plot_df_fn = \"temp_files/\" + \"plot_df_\" + \"-\".join([str(x)[0:12] for x in type_selected])+\"_\"+ str(thresh_list[0])+\"_\"+str(bam_fracs[0])+ \".csv\"\n",
    "comb_bedmethyl_plot_df.to_csv(plot_df_fn, index=False)\n",
    "if os.path.exists(plot_df_fn):\n",
    "    print(\"plot_df_fn exists, importing...\")\n",
    "    plot_df = pd.read_csv(plot_df_fn)\n",
    "else:\n",
    "    print(\"plot_df_fn does not exist, creating...\")\n",
    "    comb_bedmethyl_plot_df.to_csv(plot_df_fn, index=False)\n",
    "    plot_df = comb_bedmethyl_plot_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
