{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "608d7544",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Purpose of this script is to take in a .bam file and a .bed file and return %m6A across all regions in the .bam file.\n",
    "# This is focused on mex motifs\n",
    "__author__ = \"Yuri Malina\"\n",
    "__contact__ = \"ymalina@berkeley.edu\"\n",
    "__copyright__ = \"The Meyer Lab, UC Berkeley\"\n",
    "__credits__ = [\"\"]\n",
    "__date__ = \"2/6/2023\"\n",
    "__deprecated__ = False\n",
    "__status__ = \"In development\"\n",
    "__version__ = \"0.0.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5cd4207",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#import seaborn as sns\n",
    "import plotly.express as px\n",
    "#import plotly.graph_objects as go\n",
    "#from plotly.subplots import make_subplots\n",
    "import numpy as np\n",
    "import pysam\n",
    "import random\n",
    "from scipy import stats\n",
    "# install tabix with:\n",
    "# apt-get install tabix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1c9bc6a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Configurations\n",
    "m6A_thresh = 129 #default is 129\n",
    "mC_thresh = 129 #default is 129\n",
    "coreNum = 96 # cores to use\n",
    "\n",
    "## Bed file configurations:\n",
    "sample_source = \"type\" # \"chr_type\" or \"type\" or \"chromosome\"\n",
    "sampleName = [\"CHROMOSOME_I\", \"CHROMOSOME_II\", \"CHROMOSOME_III\", \"CHROMOSOME_IV\", \"CHROMOSOME_V\",\"CHROMOSOME_X\"] # \"TES_q1\" \"strong_rex\" \"weak_rex\" \"type\", \"X\", \"Autosome\"; Must be same number of unique values in selected bed rows.\n",
    "chr_type_selected = [\"X\",\"Autosome\"] # 'X' or \"Autosome\"\n",
    "type_selected = [\"mex_motif_strong_rex\"] #TES_q1-4 | #TSS_q1-4 | strong/weak rex | whole_chr | 200kb_region | 50kb_region | mex_motif | mexII_motif\n",
    "max_regions = 0 # max regions to consider; 0 = full set;\n",
    "chromosome_selected = [\"CHROMOSOME_I\", \"CHROMOSOME_II\", \"CHROMOSOME_III\", \"CHROMOSOME_IV\", \"CHROMOSOME_V\",\"CHROMOSOME_X\"]\n",
    "strand_selected = [\"+\"] #+ and/or -\n",
    "bed_file = \"~/Data1/reference/tss_tes_rex_combined.bed\"\n",
    "mods = \"A\" # {A,CG,A+CG}\n",
    "if sample_source == \"chr_type\":\n",
    "    selection = chr_type_selected\n",
    "if sample_source == \"type\":\n",
    "    selection = type_selected\n",
    "if sample_source == \"chromosome\":\n",
    "    selection = chromosome_selected\n",
    "\n",
    "## Bam file configurations\n",
    "random.seed(10)\n",
    "\n",
    "### Tube D\n",
    "#bam_frac = 1 # For full .bam set to = 1\n",
    "#bam_file = \"/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/basecalls/m6A/mod_mappings.sorted.bam\"\n",
    "#output_stem = \"/Data1/seq_data/TubeD1a_N2_Fiberseq_Hia5_MSssI_12_22_22/basecalls/m6A/\"\n",
    "#condition = \"N2; 2uM Hia5 30min\"\n",
    "\n",
    "### Tube 4\n",
    "#bam_frac = 1 # For full .bam set to = 1\n",
    "#bam_file = \"/Data1/seq_data/Tube4_b2_2uM-Hia5_fiber-seq_11_21_22/basecalls/mod_mappings.sorted.m6Aonly.bam\"\n",
    "#output_stem = \"/Data1/seq_data/Tube4_b2_2uM-Hia5_fiber-seq_11_21_22/basecalls/\"\n",
    "#condition = \"N2; 2uM Hia5 120min\"\n",
    "\n",
    "### Tube H\n",
    "bam_frac = 1 # For full .bam set to = 1\n",
    "bam_file = \"/Data1/seq_data/TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19/basecalls/m6A/mod_mappings.sorted.bam\"\n",
    "output_stem = \"/Data1/seq_data/TubeH1_021_SDC2-AIDpAux_Hia5_MSssI_12_19/basecalls/m6A/\"\n",
    "condition = \"AID::SDC-2 + Auxin; 2uM Hia5 30min\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c5a8a91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      chromosome     start       end strand                  type  chr-type\n",
      "0   CHROMOSOME_I   6190609   6190620      +  mex_motif_strong_rex  Autosome\n",
      "1   CHROMOSOME_I   6349634   6349645      +  mex_motif_strong_rex  Autosome\n",
      "2   CHROMOSOME_I   6837061   6837072      +  mex_motif_strong_rex  Autosome\n",
      "3   CHROMOSOME_I   7220390   7220401      +  mex_motif_strong_rex  Autosome\n",
      "4   CHROMOSOME_I   8719626   8719637      +  mex_motif_strong_rex  Autosome\n",
      "..           ...       ...       ...    ...                   ...       ...\n",
      "72  CHROMOSOME_X  12363627  12363638      +  mex_motif_strong_rex         X\n",
      "73  CHROMOSOME_X  12365162  12365173      +  mex_motif_strong_rex         X\n",
      "74  CHROMOSOME_X  12730301  12730312      +  mex_motif_strong_rex         X\n",
      "75  CHROMOSOME_X  13514556  13514567      +  mex_motif_strong_rex         X\n",
      "76  CHROMOSOME_X  14525923  14525934      +  mex_motif_strong_rex         X\n",
      "\n",
      "[77 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21356/1855817302.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  temp_bed.drop(drop_indices,inplace=True)\n",
      "/tmp/ipykernel_21356/1855817302.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  temp_bed.sort_values(by=[\"chromosome\",\"start\"],ascending=True,inplace=True)\n",
      "/tmp/ipykernel_21356/1855817302.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  temp_bed[\"start\"]=temp_bed[\"start\"]\n",
      "/tmp/ipykernel_21356/1855817302.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  temp_bed[\"end\"]=temp_bed[\"end\"]\n"
     ]
    }
   ],
   "source": [
    "### Select bed file\n",
    "full_bed = pd.read_csv(bed_file,sep='\\t')\n",
    "bed=[]\n",
    "\n",
    "for each_type in selection:\n",
    "# REGION CONFIGURATION\n",
    "    if sample_source == \"type\":\n",
    "        temp_bed = full_bed[full_bed[\"chromosome\"].isin(chromosome_selected) &\n",
    "                            full_bed[\"chr-type\"].isin(chr_type_selected) &\n",
    "                            full_bed[\"type\"].str.contains(each_type) &\n",
    "                            full_bed[\"strand\"].isin(strand_selected)]\n",
    "    if sample_source == \"chr_type\":\n",
    "        temp_bed = full_bed[full_bed[\"chromosome\"].isin(chromosome_selected) &\n",
    "                            full_bed[\"chr-type\"].str.contains(each_type) &\n",
    "                            full_bed[\"type\"].isin(type_selected) &\n",
    "                            full_bed[\"strand\"].isin(strand_selected)]\n",
    "    if sample_source == \"chromosome\":\n",
    "        temp_bed = full_bed[full_bed[\"chromosome\"].str.contains(each_type) &\n",
    "                            full_bed[\"chr-type\"].isin(chr_type_selected) &\n",
    "                            full_bed[\"type\"].isin(type_selected) &\n",
    "                            full_bed[\"strand\"].isin(strand_selected)]\n",
    "\n",
    "    # Drop random regions to match max_regions\n",
    "    drop_count = len(temp_bed)-max_regions\n",
    "    # If max regions > selected regions, do not drop any.\n",
    "    if(drop_count<0):\n",
    "        drop_count=0\n",
    "    # If max_regions = 0, do not drop any.\n",
    "    if (max_regions == 0):\n",
    "        drop_count = 0\n",
    "\n",
    "    drop_indices = np.random.choice(temp_bed.index, drop_count, replace=False)\n",
    "    temp_bed.drop(drop_indices,inplace=True)\n",
    "    temp_bed.sort_values(by=[\"chromosome\",\"start\"],ascending=True,inplace=True)\n",
    "    temp_bed.reset_index(drop=True, inplace=True)\n",
    "    temp_bed[\"start\"]=temp_bed[\"start\"]\n",
    "    temp_bed[\"end\"]=temp_bed[\"end\"]\n",
    "    temp_bedfile = \"/Data1/reference/temp_do_not_use_\"+each_type+\".bed\"\n",
    "    temp_bedfile_gz = \"/Data1/reference/temp_do_not_use_\"+each_type+\".bed.gz\"\n",
    "    temp_bed.to_csv(temp_bedfile, sep=\"\\t\",header=False,index=False)\n",
    "    \n",
    "    # Create indexed tabix files\n",
    "    print(temp_bed)\n",
    "    ! bgzip -c {temp_bedfile} > {temp_bedfile_gz}\n",
    "    ! tabix -f -p bed {temp_bedfile_gz}\n",
    "\n",
    "    # For first iteration\n",
    "    if bed == []:\n",
    "        bed = [temp_bedfile]\n",
    "\n",
    "    # Otherwise append region to temporary bed file.\n",
    "    else:\n",
    "        bed.append(temp_bedfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b79d0697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROMOSOME_I\t6190609\t6190620\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t6837061\t6837072\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t7220390\t7220401\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t8719626\t8719637\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t10384362\t10384373\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t12681339\t12681350\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t13121680\t13121691\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_I\t14162106\t14162117\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t2205625\t2205636\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t3518200\t3518211\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t4024252\t4024263\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t4574953\t4574964\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t4752783\t4752794\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t5005585\t5005596\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t5790101\t5790112\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t6383053\t6383064\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t6770504\t6770515\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t9217402\t9217413\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t10003334\t10003345\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t10417635\t10417646\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t11974069\t11974080\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t13096975\t13096986\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t13504736\t13504747\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t13577801\t13577812\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t13664840\t13664851\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t14693715\t14693726\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_II\t14711771\t14711782\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t2759791\t2759802\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t3126470\t3126481\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t4052991\t4053002\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t4185179\t4185190\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t5499685\t5499696\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t6465078\t6465089\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t7466710\t7466721\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t8488686\t8488697\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t8795911\t8795922\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t9112630\t9112641\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t10297458\t10297469\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_III\t13432355\t13432366\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t3804750\t3804761\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t7558982\t7558993\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t8193051\t8193062\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t8302125\t8302136\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t9580289\t9580300\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t14632298\t14632309\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t15518622\t15518633\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t15890841\t15890852\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t15943285\t15943296\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t15958264\t15958275\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_IV\t16498287\t16498298\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t2831197\t2831208\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t7775778\t7775789\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t8962367\t8962378\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t9789648\t9789659\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t10429981\t10429992\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t13607764\t13607775\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t14826131\t14826142\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t15119089\t15119100\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t16806559\t16806570\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_V\t17951427\t17951438\t+\tmex_motif_strong_rex\tAutosome\n",
      "CHROMOSOME_X\t806672\t806683\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t806748\t806759\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t4208737\t4208748\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t4209290\t4209301\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t5429588\t5429599\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t6296509\t6296520\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t6296546\t6296557\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t6296574\t6296585\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t8036433\t8036444\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t8048831\t8048842\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t11899364\t11899375\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t12363627\t12363638\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t12365162\t12365173\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t12730301\t12730312\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t13514556\t13514567\t+\tmex_motif_strong_rex\tX\n",
      "CHROMOSOME_X\t14525923\t14525934\t+\tmex_motif_strong_rex\tX\n"
     ]
    }
   ],
   "source": [
    "### Extract only reads in bam file that overlap selected regions, and subselect down using fraction\n",
    "# Define function to generate redux .bam file\n",
    "def extract_overlapping_reads(bed_file, bam_file, output_file):\n",
    "    # Load the BAM file\n",
    "    bam_ext = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "\n",
    "    # Load the BED file\n",
    "    regions = pysam.TabixFile(bed_file)\n",
    "\n",
    "    '''# Create the output BAM file\n",
    "    out_bam = pysam.AlignmentFile(output_file, \"wb\", template=bam)\n",
    "\n",
    "    # Iterate over the reads in the input BAM file\n",
    "    for read in bam:\n",
    "        # Check if the read overlaps a region in the BED file\n",
    "        overlaps = [r for r in regions.fetch(read.reference_name, read.reference_start, read.reference_end)]\n",
    "        if len(overlaps) > 0:\n",
    "            # Write the read to the output BAM file if it overlaps a region in the BED file\n",
    "            out_bam.write(read)\n",
    "\n",
    "    # Close the input BAM file and the output BAM file\n",
    "    bam.close()\n",
    "    out_bam.close()'''\n",
    "\n",
    "    # Create a set to store the reads\n",
    "    seen = set()\n",
    "\n",
    "    # Create the output .bam file\n",
    "    with pysam.AlignmentFile(output_file, \"wb\", header=bam_ext.header) as out:\n",
    "        # Iterate over the regions in the .bed file\n",
    "        for region in regions.fetch():\n",
    "            print(region)\n",
    "            chrom, start, end, strand, region_type, chr_type = region.split()\n",
    "            start, end = int(start), int(end)\n",
    "\n",
    "            # Iterate over the reads in the current region\n",
    "            for pileupcolumn in bam_ext.pileup(chrom, start, end):\n",
    "                for pileupread in pileupcolumn.pileups:\n",
    "                    # Check if the read has not been seen before\n",
    "                    if pileupread.alignment.query_name not in seen:\n",
    "                        # Add the read to the set of seen reads\n",
    "                        seen.add(pileupread.alignment.query_name)\n",
    "                        # Write the read to the output .bam file\n",
    "                        out.write(pileupread.alignment)\n",
    "    \n",
    "for each_type in selection:\n",
    "    temp_bedfile = \"/Data1/reference/temp_do_not_use_\"+each_type+\".bed.gz\"\n",
    "    output_bamfile = output_stem + \"mod_mappings_\" + str(bam_frac)+\"_\"+each_type+\".sorted.bam\"\n",
    "    extract_overlapping_reads(temp_bedfile, bam_file, output_bamfile)\n",
    "    ### Subselect bam file using fraction\n",
    "    ! samtools view -h -s {bam_frac} -L {temp_bedfile} {bam_file} | samtools view -h -b - > {output_bamfile}\n",
    "    ! samtools index {output_bamfile}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f63e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Caculate average m6A per region based on input bed and bam files\n",
    "def extract_m6A_per_base(bam_file, bed_file, threshold):\n",
    "    # Load the BAM file\n",
    "    bam_ext = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "\n",
    "    # Load the BED file\n",
    "    regions = pysam.TabixFile(bed_file)\n",
    "\n",
    "    # Initialize a list to store the results\n",
    "    results = []\n",
    "\n",
    "    # Iterate over the regions in the BED file\n",
    "    methylation_status = pd.DataFrame(columns=[\"chr\",\"start\",\"end\",\"strand\",\"region_type\",\"chr_type\",\n",
    "                                   \"read\",\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\"])\n",
    "    for region in regions.fetch():\n",
    "        # Split the region string into the chromosome, start, and end positions\n",
    "        #print(\"Starting on region:\",region)\n",
    "        chromosome, start, end, strand, region_type, chr_type = region.split()\n",
    "        start = int(start)\n",
    "        end = int(end)\n",
    "\n",
    "        # Initialize counters for the total number of bases and the total number of m6A\n",
    "        total_bases = 0\n",
    "        total_m6A = 0\n",
    "        # Iterate over the reads that overlap the region\n",
    "        \n",
    "        for read in bam_ext.fetch(chromosome, start, end):\n",
    "            if not read.is_unmapped and read.is_forward == True:\n",
    "                    read_name = read.query_name\n",
    "                    seq_row = [chromosome, start, end, strand, region_type, chr_type,read_name]\n",
    "                    total_A = 0\n",
    "                    m6A_dict = read.modified_bases[('A', 0, 'Y')] #('A', 0, 'a')\n",
    "                    #print(m6A_dict)\n",
    "                    \n",
    "                    for (i, base) in enumerate(read.seq):\n",
    "                        #print(\"start:\",start,\" | read.reference_start + i\",read.reference_start + i,\" | end:\",end)\n",
    "                        if base == \"A\":\n",
    "                            total_A += 1\n",
    "                        if start-1 <= read.reference_start + i < end:\n",
    "                            #print(\"i,base\",i,\" , \", base)\n",
    "                            if base == \"A\":\n",
    "                                #print(\"i: \",i)\n",
    "                                try: base = [item for item in m6A_dict if item[0] == i][0][1]\n",
    "                                except: base = 0\n",
    "                            else:\n",
    "                                base = np.NaN\n",
    "                            seq_row.append(base)\n",
    "                            #print(\"seq_row\",seq_row)\n",
    "                    #print(\"seq_row:\",seq_row)\n",
    "                    if len(seq_row) == len(methylation_status.columns):\n",
    "                        methylation_status.loc[len(methylation_status)+1] = seq_row\n",
    "    print(\"METH STATUS=\",methylation_status)\n",
    "    return methylation_status\n",
    "\n",
    "for each_type in selection:\n",
    "    bam = output_stem + \"mod_mappings_\" + str(bam_frac)+\"_\"+each_type+\".sorted.bam\"\n",
    "    temp_bedfile = \"/Data1/reference/temp_do_not_use_\"+each_type+\".bed.gz\"\n",
    "    output_df = extract_m6A_per_base(bam,temp_bedfile,m6A_thresh)\n",
    "    #output_df.to_csv(output_stem + \"m6A_frac\" + str(bam_frac)+\"_\"+each_type+\".csv\", index=False, mode='w')\n",
    "    print(output_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5fc188a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#source: https://stackoverflow.com/questions/67505252/plotly-box-p-value-significant-annotation\n",
    "def add_p_value_annotation(fig, array_columns, subplot=None, _format=dict(interline=0.07, text_height=1.07, color='black')):\n",
    "    ''' Adds notations giving the p-value between two box plot data (t-test two-sided comparison)\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    fig: figure\n",
    "        plotly boxplot figure\n",
    "    array_columns: np.array\n",
    "        array of which columns to compare \n",
    "        e.g.: [[0,1], [1,2]] compares column 0 with 1 and 1 with 2\n",
    "    subplot: None or int\n",
    "        specifies if the figures has subplots and what subplot to add the notation to\n",
    "    _format: dict\n",
    "        format characteristics for the lines\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    fig: figure\n",
    "        figure with the added notation\n",
    "    '''\n",
    "    # Specify in what y_range to plot for each pair of columns\n",
    "    y_range = np.zeros([len(array_columns), 2])\n",
    "    for i in range(len(array_columns)):\n",
    "        y_range[i] = [1.01+i*_format['interline'], 1.02+i*_format['interline']]\n",
    "\n",
    "    # Get values from figure\n",
    "    fig_dict = fig.to_dict()\n",
    "\n",
    "    # Get indices if working with subplots\n",
    "    if subplot:\n",
    "        if subplot == 1:\n",
    "            subplot_str = ''\n",
    "        else:\n",
    "            subplot_str =str(subplot)\n",
    "        indices = [] #Change the box index to the indices of the data for that subplot\n",
    "        for index, data in enumerate(fig_dict['data']):\n",
    "            #print(index, data['xaxis'], 'x' + subplot_str)\n",
    "            if data['xaxis'] == 'x' + subplot_str:\n",
    "                indices = np.append(indices, index)\n",
    "        indices = [int(i) for i in indices]\n",
    "        print((indices))\n",
    "    else:\n",
    "        subplot_str = ''\n",
    "\n",
    "    # Print the p-values\n",
    "    for index, column_pair in enumerate(array_columns):\n",
    "        if subplot:\n",
    "            data_pair = [indices[column_pair[0]], indices[column_pair[1]]]\n",
    "        else:\n",
    "            data_pair = column_pair\n",
    "\n",
    "        # Mare sure it is selecting the data and subplot you want\n",
    "        #print('0:', fig_dict['data'][data_pair[0]]['name'], fig_dict['data'][data_pair[0]]['xaxis'])\n",
    "        #print('1:', fig_dict['data'][data_pair[1]]['name'], fig_dict['data'][data_pair[1]]['xaxis'])\n",
    "\n",
    "        # Get the p-value\n",
    "        pvalue = stats.ttest_ind(\n",
    "            fig_dict['data'][data_pair[0]]['y'],\n",
    "            fig_dict['data'][data_pair[1]]['y'],\n",
    "            equal_var=False,\n",
    "        )[1]\n",
    "        if pvalue >= 0.05:\n",
    "            symbol = 'ns'\n",
    "        elif pvalue >= 0.01: \n",
    "            symbol = '*'\n",
    "        elif pvalue >= 0.001:\n",
    "            symbol = '**'\n",
    "        else:\n",
    "            symbol = '***'\n",
    "        # Vertical line\n",
    "        fig.add_shape(type=\"line\",\n",
    "            xref=\"x\"+subplot_str, yref=\"y\"+subplot_str+\" domain\",\n",
    "            x0=column_pair[0], y0=y_range[index][0], \n",
    "            x1=column_pair[0], y1=y_range[index][1],\n",
    "            line=dict(color=_format['color'], width=2,)\n",
    "        )\n",
    "        # Horizontal line\n",
    "        fig.add_shape(type=\"line\",\n",
    "            xref=\"x\"+subplot_str, yref=\"y\"+subplot_str+\" domain\",\n",
    "            x0=column_pair[0], y0=y_range[index][1], \n",
    "            x1=column_pair[1], y1=y_range[index][1],\n",
    "            line=dict(color=_format['color'], width=2,)\n",
    "        )\n",
    "        # Vertical line\n",
    "        fig.add_shape(type=\"line\",\n",
    "            xref=\"x\"+subplot_str, yref=\"y\"+subplot_str+\" domain\",\n",
    "            x0=column_pair[1], y0=y_range[index][0], \n",
    "            x1=column_pair[1], y1=y_range[index][1],\n",
    "            line=dict(color=_format['color'], width=2,)\n",
    "        )\n",
    "        ## add text at the correct x, y coordinates\n",
    "        ## for bars, there is a direct mapping from the bar number to 0, 1, 2...\n",
    "        fig.add_annotation(dict(font=dict(color=_format['color'],size=14),\n",
    "            x=(column_pair[0] + column_pair[1])/2,\n",
    "            y=y_range[index][1]*_format['text_height'],\n",
    "            showarrow=False,\n",
    "            text=symbol,\n",
    "            textangle=0,\n",
    "            xref=\"x\"+subplot_str,\n",
    "            yref=\"y\"+subplot_str+\" domain\"\n",
    "        ))\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f0fe6d5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'combined_regions' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Plot the boxplot\u001b[39;00m\n\u001b[1;32m      2\u001b[0m marker_colors \u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m#c45746\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m#16415e\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m----> 4\u001b[0m tube4_df \u001b[38;5;241m=\u001b[39m \u001b[43mcombined_regions\u001b[49m\u001b[38;5;241m.\u001b[39mloc[combined_regions[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcondition\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mN2; 2uM Hia5 120min\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m      5\u001b[0m tubeD_df \u001b[38;5;241m=\u001b[39m combined_regions\u001b[38;5;241m.\u001b[39mloc[combined_regions[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcondition\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mN2; 2uM Hia5 30min\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m      6\u001b[0m tubeH_df \u001b[38;5;241m=\u001b[39m combined_regions\u001b[38;5;241m.\u001b[39mloc[combined_regions[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcondition\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAID::SDC-2 + Auxin; 2uM Hia5 30min\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'combined_regions' is not defined"
     ]
    }
   ],
   "source": [
    "# Plot the boxplot\n",
    "marker_colors =[\"#c45746\",\"#16415e\"]\n",
    "\n",
    "tube4_df = combined_regions.loc[combined_regions['condition']==\"N2; 2uM Hia5 120min\"]\n",
    "tubeD_df = combined_regions.loc[combined_regions['condition']==\"N2; 2uM Hia5 30min\"]\n",
    "tubeH_df = combined_regions.loc[combined_regions['condition']==\"AID::SDC-2 + Auxin; 2uM Hia5 30min\"]\n",
    "\n",
    "#fig=go.Figure()\n",
    "# Source for subplot schema: https://stackoverflow.com/questions/55698429/different-box-plot-series-traces-within-plotly-subplots\n",
    "# Tube 4\n",
    "chr_type = \"Autosome\"\n",
    "df_plot=tube4_df.loc[tube4_df['chr_type']==chr_type]\n",
    "trace0 = go.Box(x=df_plot['condition']+\" \", y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[0])\n",
    "\n",
    "chr_type = \"X\"\n",
    "df_plot=tube4_df.loc[tube4_df['chr_type']==chr_type]\n",
    "trace1 = go.Box(x=df_plot['condition'], y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[1])\n",
    "\n",
    "# Tube D\n",
    "chr_type = \"Autosome\"\n",
    "df_plot=tubeD_df.loc[tubeD_df['chr_type']==chr_type]\n",
    "trace2 = go.Box(x=df_plot['condition']+\" \", y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[0])\n",
    "\n",
    "chr_type = \"X\"\n",
    "df_plot=tubeD_df.loc[tubeD_df['chr_type']==chr_type]\n",
    "trace3 = go.Box(x=df_plot['condition'], y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[1])\n",
    "\n",
    "# Tube H\n",
    "chr_type = \"Autosome\"\n",
    "df_plot=tubeH_df.loc[tubeH_df['chr_type']==chr_type]\n",
    "trace4 = go.Box(x=df_plot['condition']+\" \", y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[0])\n",
    "\n",
    "chr_type = \"X\"\n",
    "df_plot=tubeH_df.loc[tubeH_df['chr_type']==chr_type]\n",
    "trace5 = go.Box(x=df_plot['condition'], y=df_plot['norm_m6A_frac'],\n",
    "                     notched=True,\n",
    "                     name=chr_type, marker_color =marker_colors[1])\n",
    "\n",
    "fig = make_subplots(rows=1, cols=3,\n",
    "                    y_title = \"m6A/A (normalized to average Autosome)\",\n",
    "                    shared_yaxes=True,\n",
    "                    subplot_titles=(\"N2; <br> 2uM Hia5 120min\", \n",
    "                                    \"N2; <br> 2uM Hia5 30min\", \n",
    "                                    \"AID::SDC-2 + Auxin; <br> 2uM Hia5 30min\"))\n",
    "fig.append_trace(trace0, row = 1, col = 1)\n",
    "fig.append_trace(trace1, row = 1, col = 1)\n",
    "fig.append_trace(trace2, row = 1, col = 2)\n",
    "fig.append_trace(trace3, row = 1, col = 2)\n",
    "fig.append_trace(trace4, row = 1, col = 3)\n",
    "fig.append_trace(trace5, row = 1, col = 3)\n",
    "fig = add_p_value_annotation(fig, [[0,1]],0)\n",
    "fig = add_p_value_annotation(fig, [[0,1]],2)\n",
    "fig = add_p_value_annotation(fig, [[0,1]],3)\n",
    "fig.update_yaxes(dtick=0.05)\n",
    "fig.layout.annotations[0].update(y=-0.1)\n",
    "fig.layout.annotations[2].update(y=-0.1)\n",
    "fig.layout.annotations[1].update(y=-0.1)\n",
    "fig['layout'].update(height = 600)\n",
    "fig.update_layout(template=\"plotly_white\")\n",
    "fig.update_xaxes(showticklabels=False)\n",
    "    \n",
    "#fig.update_layout(boxmode='group', xaxis_tickangle=0)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fd3905",
   "metadata": {},
   "outputs": [],
   "source": [
    "#output_df_mean = output_df[[\"chr_type\",\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\"]].groupby('chr_type',as_index=False).mean().transpose()\n",
    "output_df_melt = output_df.melt(id_vars=[\"chr_type\"],value_vars=[\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\"]).dropna()\n",
    "print(output_df_melt)\n",
    "output_df_melt['m6A_binary']=np.where(output_df_melt['value'] > m6A_thresh, True, False)\n",
    "#print(output_df_melt)\n",
    "output_df_count = output_df_melt.groupby(['variable','chr_type']).size().reset_index(name='count')\n",
    "output_df_count['variable']=output_df_count['variable'].astype(int)\n",
    "output_df_count=output_df_count.sort_values(by=[\"chr_type\",\"variable\"],ascending=True)\n",
    "output_df_count['variable']=output_df_count['variable'].astype(str)\n",
    "output_df_count['percent']=output_df_count['count']\n",
    "def calculate_percent(row):\n",
    "    if row['chr_type'] == 'Autosome':\n",
    "        return row['count']/output_df['chr_type'].value_counts()['Autosome']\n",
    "    elif row['chr_type'] == 'X':\n",
    "        return row['count']/output_df['chr_type'].value_counts()['X']\n",
    "\n",
    "output_df_count['percent'] = output_df_count.apply(calculate_percent, axis=1)\n",
    "\n",
    "\n",
    "print(output_df_count)\n",
    "output_df_mean = output_df_melt.groupby(['variable','chr_type']).mean().reset_index()\n",
    "output_df_mean['variable']=output_df_mean['variable'].astype(int)\n",
    "output_df_mean=output_df_mean.sort_values(by=[\"chr_type\",\"variable\"],ascending=True)\n",
    "output_df_mean['variable']=output_df_mean['variable'].astype(str)\n",
    "#print(output_df_mean)\n",
    "#print(output_df_melt)\n",
    "\n",
    "# Plot the boxplot\n",
    "marker_colors =[\"#c45746\",\"#16415e\"]\n",
    "\n",
    "fig = px.box(output_df_melt, x=\"variable\", y=\"value\", color=\"chr_type\")\n",
    "fig2 = px.bar(output_df_mean, x=\"variable\", y=\"m6A_binary\", color=\"chr_type\",barmode=\"group\")\n",
    "fig3 = px.bar(output_df_count, x=\"variable\", y=\"percent\", color=\"chr_type\",barmode=\"group\")\n",
    "fig2.update_layout(template=\"plotly_white\")\n",
    "fig3.update_layout(template=\"plotly_white\")\n",
    "\n",
    "#fig.show()\n",
    "fig2.show()\n",
    "fig3.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
